{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c3f2875f-bb35-4dfe-a585-92cf55c6f7d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“package ‘ggrepel’ was built under R version 3.6.3”\n",
      "Warning message:\n",
      "“package ‘ggseqlogo’ was built under R version 3.6.3”\n",
      "Warning message:\n",
      "“package ‘pheatmap’ was built under R version 3.6.3”\n",
      "Joining, by = \"species\"\n",
      "\n"
     ]
    }
   ],
   "source": [
    "source(file.path(Sys.getenv(\"CODEBASE\"),\"DNAmeth500species/src/00.0_init.R\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d5b4ebc6-d92c-4401-b816-e845337efc46",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_dir = \"../../resources/reference_genomes/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4c7ccedc-a8cc-4c02-adc1-2e5c7d428c00",
   "metadata": {},
   "outputs": [],
   "source": [
    "genomes_df <- fread(\"/binfl/lv71484/droman/DNAmeth500species/results_analysis/validation/02_insilico_digest/genomes_to_run.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4117266b-0606-4344-ad0c-e901ce055983",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>V2</th><th scope=col>V3</th><th scope=col>V4</th><th scope=col>V5</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>vicPac2</td><td>Alpaca      </td><td>NA          </td><td>NA                     </td><td>Mammalia    </td></tr>\n",
       "\t<tr><td>dasNov3</td><td>Armadillo   </td><td>NA          </td><td>NA                     </td><td>Malacostraca</td></tr>\n",
       "\t<tr><td>gasAcu1</td><td>Stickleback </td><td>NA          </td><td>NA                     </td><td>Actinopteri </td></tr>\n",
       "\t<tr><td>anoCar2</td><td>Lizard      </td><td>Lizard      </td><td>Anolis carolinensis    </td><td>Reptilia    </td></tr>\n",
       "\t<tr><td>melUnd1</td><td>Budgerigar  </td><td>Budgeriger  </td><td>Melopsittacus undulatus</td><td>Aves        </td></tr>\n",
       "\t<tr><td>aquChr2</td><td>Golden eagle</td><td>Golden eagle</td><td>Aquila chrysaetos      </td><td>Aves        </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " V1 & V2 & V3 & V4 & V5\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t vicPac2 & Alpaca       & NA           & NA                      & Mammalia    \\\\\n",
       "\t dasNov3 & Armadillo    & NA           & NA                      & Malacostraca\\\\\n",
       "\t gasAcu1 & Stickleback  & NA           & NA                      & Actinopteri \\\\\n",
       "\t anoCar2 & Lizard       & Lizard       & Anolis carolinensis     & Reptilia    \\\\\n",
       "\t melUnd1 & Budgerigar   & Budgeriger   & Melopsittacus undulatus & Aves        \\\\\n",
       "\t aquChr2 & Golden eagle & Golden eagle & Aquila chrysaetos       & Aves        \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 5\n",
       "\n",
       "| V1 &lt;chr&gt; | V2 &lt;chr&gt; | V3 &lt;chr&gt; | V4 &lt;chr&gt; | V5 &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| vicPac2 | Alpaca       | NA           | NA                      | Mammalia     |\n",
       "| dasNov3 | Armadillo    | NA           | NA                      | Malacostraca |\n",
       "| gasAcu1 | Stickleback  | NA           | NA                      | Actinopteri  |\n",
       "| anoCar2 | Lizard       | Lizard       | Anolis carolinensis     | Reptilia     |\n",
       "| melUnd1 | Budgerigar   | Budgeriger   | Melopsittacus undulatus | Aves         |\n",
       "| aquChr2 | Golden eagle | Golden eagle | Aquila chrysaetos       | Aves         |\n",
       "\n"
      ],
      "text/plain": [
       "  V1      V2           V3           V4                      V5          \n",
       "1 vicPac2 Alpaca       NA           NA                      Mammalia    \n",
       "2 dasNov3 Armadillo    NA           NA                      Malacostraca\n",
       "3 gasAcu1 Stickleback  NA           NA                      Actinopteri \n",
       "4 anoCar2 Lizard       Lizard       Anolis carolinensis     Reptilia    \n",
       "5 melUnd1 Budgerigar   Budgeriger   Melopsittacus undulatus Aves        \n",
       "6 aquChr2 Golden eagle Golden eagle Aquila chrysaetos       Aves        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(genomes_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9aa10f20-038e-496c-af3d-eb89e9ec3233",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 77 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>V2</th><th scope=col>V3</th><th scope=col>V4</th><th scope=col>V5</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>vicPac2 </td><td>Alpaca            </td><td>NA                </td><td>NA                             </td><td>Mammalia      </td></tr>\n",
       "\t<tr><td>dasNov3 </td><td>Armadillo         </td><td>NA                </td><td>NA                             </td><td>Malacostraca  </td></tr>\n",
       "\t<tr><td>gasAcu1 </td><td>Stickleback       </td><td>NA                </td><td>NA                             </td><td>Actinopteri   </td></tr>\n",
       "\t<tr><td>anoCar2 </td><td>Lizard            </td><td>Lizard            </td><td>Anolis carolinensis            </td><td>Reptilia      </td></tr>\n",
       "\t<tr><td>melUnd1 </td><td>Budgerigar        </td><td>Budgeriger        </td><td>Melopsittacus undulatus        </td><td>Aves          </td></tr>\n",
       "\t<tr><td>aquChr2 </td><td>Golden eagle      </td><td>Golden eagle      </td><td>Aquila chrysaetos              </td><td>Aves          </td></tr>\n",
       "\t<tr><td>chrPic1 </td><td>Painted turtle    </td><td>Painted turtle    </td><td>Chrysemys picta                </td><td>Reptilia      </td></tr>\n",
       "\t<tr><td>calJac4 </td><td>Marmoset          </td><td>NA                </td><td>NA                             </td><td>Mammalia      </td></tr>\n",
       "\t<tr><td>choHof1 </td><td>Sloth             </td><td>NA                </td><td>NA                             </td><td>Mammalia      </td></tr>\n",
       "\t<tr><td>saiBol1 </td><td>Squirrel monkey   </td><td>Squirrel monkey   </td><td>Saimiri boliviensis boliviensis</td><td>Mammalia      </td></tr>\n",
       "\t<tr><td>pteVam1 </td><td>Megabat           </td><td>Megabat           </td><td>Pteropus vampyrus              </td><td>Mammalia      </td></tr>\n",
       "\t<tr><td>aptMan1 </td><td>Brown kiwi        </td><td>NA                </td><td>NA                             </td><td>Aves          </td></tr>\n",
       "\t<tr><td>dipOrd1 </td><td>Kangaroo rat      </td><td>NA                </td><td>NA                             </td><td>Mammalia      </td></tr>\n",
       "\t<tr><td>oviAri4 </td><td>Sheep             </td><td>Sheep             </td><td>Ovis aries                     </td><td>Mammalia      </td></tr>\n",
       "\t<tr><td>allMis1 </td><td>American alligator</td><td>American alligator</td><td>Alligator mississippiensis     </td><td>Reptilia      </td></tr>\n",
       "\t<tr><td>canFam5 </td><td>Dog               </td><td>NA                </td><td>NA                             </td><td>Mammalia      </td></tr>\n",
       "\t<tr><td>felCat9 </td><td>Cat               </td><td>Cat               </td><td>Felis catus                    </td><td>Mammalia      </td></tr>\n",
       "\t<tr><td>susScr11</td><td>Pig               </td><td>Pig               </td><td>Sus scrofa                     </td><td>Mammalia      </td></tr>\n",
       "\t<tr><td>thaSir1 </td><td>Garter snake      </td><td>Garter snake      </td><td>Thamnophis sirtalis            </td><td>Reptilia      </td></tr>\n",
       "\t<tr><td>musFur1 </td><td>Ferret            </td><td>Ferret            </td><td>Mustela putorius furo          </td><td>Mammalia      </td></tr>\n",
       "\t<tr><td>calMil1 </td><td>Elephant shark    </td><td>Elephant shark    </td><td>Callorhinchus milii            </td><td>Chondrichthyes</td></tr>\n",
       "\t<tr><td>equCab3 </td><td>Horse             </td><td>NA                </td><td>NA                             </td><td>Mammalia      </td></tr>\n",
       "\t<tr><td>speTri2 </td><td>Squirrel          </td><td>Squirrel          </td><td>Ictidomys tridecemlineatus     </td><td>Mammalia      </td></tr>\n",
       "\t<tr><td>oryCun2 </td><td>Rabbit            </td><td>Rabbit            </td><td>Oryctolagus cuniculus          </td><td>Mammalia      </td></tr>\n",
       "\t<tr><td>cavPor3 </td><td>Guinea pig        </td><td>NA                </td><td>NA                             </td><td>Mammalia      </td></tr>\n",
       "\t<tr><td>tupBel1 </td><td>Tree shrew        </td><td>Tree shrew        </td><td>Tupaia belangeri               </td><td>Mammalia      </td></tr>\n",
       "\t<tr><td>taeGut2 </td><td>Zebra finch       </td><td>Zebra finch       </td><td>Taeniopygia guttata            </td><td>Aves          </td></tr>\n",
       "\t<tr><td>myoLuc2 </td><td>Microbat          </td><td>Microbat          </td><td>Myotis lucifugus               </td><td>Mammalia      </td></tr>\n",
       "\t<tr><td>mm39    </td><td>Mouse             </td><td>NA                </td><td>NA                             </td><td>Mammalia      </td></tr>\n",
       "\t<tr><td>nanPar1 </td><td>Tibetan frog      </td><td>Tibetan frog      </td><td>Nanorana parkeri               </td><td>Amphibia      </td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>oreNil2      </td><td>Nile tilapia            </td><td>Nile tilapia        </td><td>Oreochromis niloticus</td><td>Actinopteri </td></tr>\n",
       "\t<tr><td>danRer11     </td><td>Zebrafish               </td><td>Zebrafish           </td><td>Danio rerio          </td><td>Actinopteri </td></tr>\n",
       "\t<tr><td>petMar3      </td><td>Lamprey                 </td><td>NA                  </td><td>NA                   </td><td>Hyperoartia </td></tr>\n",
       "\t<tr><td>fr3          </td><td>Fugu                    </td><td>Fugu                </td><td>Takifugu rubripes    </td><td>Actinopteri </td></tr>\n",
       "\t<tr><td>tetNig2      </td><td>Tetraodon               </td><td>NA                  </td><td>NA                   </td><td>Actinopteri </td></tr>\n",
       "\t<tr><td>geoFor1      </td><td>Medium ground finch     </td><td>Medium ground finch </td><td>Geospiza fortis      </td><td>Aves        </td></tr>\n",
       "\t<tr><td>melGal5      </td><td>Turkey                  </td><td>NA                  </td><td>NA                   </td><td>Aves        </td></tr>\n",
       "\t<tr><td>triMan1      </td><td>Manatee                 </td><td>NA                  </td><td>NA                   </td><td>Mammalia    </td></tr>\n",
       "\t<tr><td>proCap1      </td><td>Rock hyrax              </td><td>NA                  </td><td>NA                   </td><td>Mammalia    </td></tr>\n",
       "\t<tr><td>echTel2      </td><td>Tenrec                  </td><td>NA                  </td><td>NA                   </td><td>Mammalia    </td></tr>\n",
       "\t<tr><td>macFas5      </td><td>Crab-eating macaque     </td><td>NA                  </td><td>NA                   </td><td>Mammalia    </td></tr>\n",
       "\t<tr><td>chlSab2      </td><td>Green Monkey            </td><td>Green Monkey        </td><td>Chlorocebus sabaeus  </td><td>Mammalia    </td></tr>\n",
       "\t<tr><td>rheMac10     </td><td>Rhesus                  </td><td>Rhesus              </td><td>Macaca mulatta       </td><td>Mammalia    </td></tr>\n",
       "\t<tr><td>xenLae2      </td><td>African clawed frog     </td><td>African clawed frog </td><td>Xenopus laevis       </td><td>Amphibia    </td></tr>\n",
       "\t<tr><td>xenTro9      </td><td>X. tropicalis           </td><td>Tropical clawed frog</td><td>Xenopus tropicalis   </td><td>Amphibia    </td></tr>\n",
       "\t<tr><td>rhiRox1      </td><td>Golden snub-nosed monkey</td><td>NA                  </td><td>NA                   </td><td>Mammalia    </td></tr>\n",
       "\t<tr><td>nasLar1      </td><td>Proboscis Monkey        </td><td>NA                  </td><td>NA                   </td><td>Mammalia    </td></tr>\n",
       "\t<tr><td>hg38         </td><td>Human                   </td><td>Human               </td><td>Homo sapiens         </td><td>Mammalia    </td></tr>\n",
       "\t<tr><td>braFlo1      </td><td>Lancelet                </td><td>Lancelet            </td><td>Cephalochordata      </td><td>Invertebrata</td></tr>\n",
       "\t<tr><td>aplCal1      </td><td>Sea Hare                </td><td>Sea Hare            </td><td>Aplysia californica  </td><td>Invertebrata</td></tr>\n",
       "\t<tr><td>latCha1      </td><td>Coelacanth              </td><td>Coelacanth          </td><td>Latimeria chalumnae  </td><td>Actinopteri </td></tr>\n",
       "\t<tr><td>sarHar1      </td><td>Tasmanian devil         </td><td>Tasmanian devil     </td><td>Sarcophilus harrisii </td><td>Marsupiala  </td></tr>\n",
       "\t<tr><td>monDom5      </td><td>Opossum                 </td><td>Opossum             </td><td>Monodelphis domestica</td><td>Marsupiala  </td></tr>\n",
       "\t<tr><td>nomLeu3      </td><td>Gibbon                  </td><td>Gibbon              </td><td>Nomascus leucogenys  </td><td>Mammalia    </td></tr>\n",
       "\t<tr><td>gorGor6      </td><td>Gorilla                 </td><td>Gorilla             </td><td>Gorilla gorilla      </td><td>Mammalia    </td></tr>\n",
       "\t<tr><td>panTro6      </td><td>Chimp                   </td><td>Chimp               </td><td>Pan troglodytes      </td><td>Mammalia    </td></tr>\n",
       "\t<tr><td>panPan2      </td><td>Bonobo                  </td><td>Bonobo              </td><td>Pan paniscus         </td><td>Mammalia    </td></tr>\n",
       "\t<tr><td>sorAra2      </td><td>Shrew                   </td><td>Shrew               </td><td>Sorex araneus        </td><td>Mammalia    </td></tr>\n",
       "\t<tr><td>macEug2      </td><td>Wallaby                 </td><td>Wallaby             </td><td>Notamacropus eugenii </td><td>Marsupiala  </td></tr>\n",
       "\t<tr><td>AmexG_v6.0-DD</td><td>Axolotl                 </td><td>Axolotl             </td><td>Ambystoma mexicanum  </td><td>Amphibia    </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 77 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " V1 & V2 & V3 & V4 & V5\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t vicPac2  & Alpaca             & NA                 & NA                              & Mammalia      \\\\\n",
       "\t dasNov3  & Armadillo          & NA                 & NA                              & Malacostraca  \\\\\n",
       "\t gasAcu1  & Stickleback        & NA                 & NA                              & Actinopteri   \\\\\n",
       "\t anoCar2  & Lizard             & Lizard             & Anolis carolinensis             & Reptilia      \\\\\n",
       "\t melUnd1  & Budgerigar         & Budgeriger         & Melopsittacus undulatus         & Aves          \\\\\n",
       "\t aquChr2  & Golden eagle       & Golden eagle       & Aquila chrysaetos               & Aves          \\\\\n",
       "\t chrPic1  & Painted turtle     & Painted turtle     & Chrysemys picta                 & Reptilia      \\\\\n",
       "\t calJac4  & Marmoset           & NA                 & NA                              & Mammalia      \\\\\n",
       "\t choHof1  & Sloth              & NA                 & NA                              & Mammalia      \\\\\n",
       "\t saiBol1  & Squirrel monkey    & Squirrel monkey    & Saimiri boliviensis boliviensis & Mammalia      \\\\\n",
       "\t pteVam1  & Megabat            & Megabat            & Pteropus vampyrus               & Mammalia      \\\\\n",
       "\t aptMan1  & Brown kiwi         & NA                 & NA                              & Aves          \\\\\n",
       "\t dipOrd1  & Kangaroo rat       & NA                 & NA                              & Mammalia      \\\\\n",
       "\t oviAri4  & Sheep              & Sheep              & Ovis aries                      & Mammalia      \\\\\n",
       "\t allMis1  & American alligator & American alligator & Alligator mississippiensis      & Reptilia      \\\\\n",
       "\t canFam5  & Dog                & NA                 & NA                              & Mammalia      \\\\\n",
       "\t felCat9  & Cat                & Cat                & Felis catus                     & Mammalia      \\\\\n",
       "\t susScr11 & Pig                & Pig                & Sus scrofa                      & Mammalia      \\\\\n",
       "\t thaSir1  & Garter snake       & Garter snake       & Thamnophis sirtalis             & Reptilia      \\\\\n",
       "\t musFur1  & Ferret             & Ferret             & Mustela putorius furo           & Mammalia      \\\\\n",
       "\t calMil1  & Elephant shark     & Elephant shark     & Callorhinchus milii             & Chondrichthyes\\\\\n",
       "\t equCab3  & Horse              & NA                 & NA                              & Mammalia      \\\\\n",
       "\t speTri2  & Squirrel           & Squirrel           & Ictidomys tridecemlineatus      & Mammalia      \\\\\n",
       "\t oryCun2  & Rabbit             & Rabbit             & Oryctolagus cuniculus           & Mammalia      \\\\\n",
       "\t cavPor3  & Guinea pig         & NA                 & NA                              & Mammalia      \\\\\n",
       "\t tupBel1  & Tree shrew         & Tree shrew         & Tupaia belangeri                & Mammalia      \\\\\n",
       "\t taeGut2  & Zebra finch        & Zebra finch        & Taeniopygia guttata             & Aves          \\\\\n",
       "\t myoLuc2  & Microbat           & Microbat           & Myotis lucifugus                & Mammalia      \\\\\n",
       "\t mm39     & Mouse              & NA                 & NA                              & Mammalia      \\\\\n",
       "\t nanPar1  & Tibetan frog       & Tibetan frog       & Nanorana parkeri                & Amphibia      \\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t oreNil2       & Nile tilapia             & Nile tilapia         & Oreochromis niloticus & Actinopteri \\\\\n",
       "\t danRer11      & Zebrafish                & Zebrafish            & Danio rerio           & Actinopteri \\\\\n",
       "\t petMar3       & Lamprey                  & NA                   & NA                    & Hyperoartia \\\\\n",
       "\t fr3           & Fugu                     & Fugu                 & Takifugu rubripes     & Actinopteri \\\\\n",
       "\t tetNig2       & Tetraodon                & NA                   & NA                    & Actinopteri \\\\\n",
       "\t geoFor1       & Medium ground finch      & Medium ground finch  & Geospiza fortis       & Aves        \\\\\n",
       "\t melGal5       & Turkey                   & NA                   & NA                    & Aves        \\\\\n",
       "\t triMan1       & Manatee                  & NA                   & NA                    & Mammalia    \\\\\n",
       "\t proCap1       & Rock hyrax               & NA                   & NA                    & Mammalia    \\\\\n",
       "\t echTel2       & Tenrec                   & NA                   & NA                    & Mammalia    \\\\\n",
       "\t macFas5       & Crab-eating macaque      & NA                   & NA                    & Mammalia    \\\\\n",
       "\t chlSab2       & Green Monkey             & Green Monkey         & Chlorocebus sabaeus   & Mammalia    \\\\\n",
       "\t rheMac10      & Rhesus                   & Rhesus               & Macaca mulatta        & Mammalia    \\\\\n",
       "\t xenLae2       & African clawed frog      & African clawed frog  & Xenopus laevis        & Amphibia    \\\\\n",
       "\t xenTro9       & X. tropicalis            & Tropical clawed frog & Xenopus tropicalis    & Amphibia    \\\\\n",
       "\t rhiRox1       & Golden snub-nosed monkey & NA                   & NA                    & Mammalia    \\\\\n",
       "\t nasLar1       & Proboscis Monkey         & NA                   & NA                    & Mammalia    \\\\\n",
       "\t hg38          & Human                    & Human                & Homo sapiens          & Mammalia    \\\\\n",
       "\t braFlo1       & Lancelet                 & Lancelet             & Cephalochordata       & Invertebrata\\\\\n",
       "\t aplCal1       & Sea Hare                 & Sea Hare             & Aplysia californica   & Invertebrata\\\\\n",
       "\t latCha1       & Coelacanth               & Coelacanth           & Latimeria chalumnae   & Actinopteri \\\\\n",
       "\t sarHar1       & Tasmanian devil          & Tasmanian devil      & Sarcophilus harrisii  & Marsupiala  \\\\\n",
       "\t monDom5       & Opossum                  & Opossum              & Monodelphis domestica & Marsupiala  \\\\\n",
       "\t nomLeu3       & Gibbon                   & Gibbon               & Nomascus leucogenys   & Mammalia    \\\\\n",
       "\t gorGor6       & Gorilla                  & Gorilla              & Gorilla gorilla       & Mammalia    \\\\\n",
       "\t panTro6       & Chimp                    & Chimp                & Pan troglodytes       & Mammalia    \\\\\n",
       "\t panPan2       & Bonobo                   & Bonobo               & Pan paniscus          & Mammalia    \\\\\n",
       "\t sorAra2       & Shrew                    & Shrew                & Sorex araneus         & Mammalia    \\\\\n",
       "\t macEug2       & Wallaby                  & Wallaby              & Notamacropus eugenii  & Marsupiala  \\\\\n",
       "\t AmexG\\_v6.0-DD & Axolotl                  & Axolotl              & Ambystoma mexicanum   & Amphibia    \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 77 × 5\n",
       "\n",
       "| V1 &lt;chr&gt; | V2 &lt;chr&gt; | V3 &lt;chr&gt; | V4 &lt;chr&gt; | V5 &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| vicPac2  | Alpaca             | NA                 | NA                              | Mammalia       |\n",
       "| dasNov3  | Armadillo          | NA                 | NA                              | Malacostraca   |\n",
       "| gasAcu1  | Stickleback        | NA                 | NA                              | Actinopteri    |\n",
       "| anoCar2  | Lizard             | Lizard             | Anolis carolinensis             | Reptilia       |\n",
       "| melUnd1  | Budgerigar         | Budgeriger         | Melopsittacus undulatus         | Aves           |\n",
       "| aquChr2  | Golden eagle       | Golden eagle       | Aquila chrysaetos               | Aves           |\n",
       "| chrPic1  | Painted turtle     | Painted turtle     | Chrysemys picta                 | Reptilia       |\n",
       "| calJac4  | Marmoset           | NA                 | NA                              | Mammalia       |\n",
       "| choHof1  | Sloth              | NA                 | NA                              | Mammalia       |\n",
       "| saiBol1  | Squirrel monkey    | Squirrel monkey    | Saimiri boliviensis boliviensis | Mammalia       |\n",
       "| pteVam1  | Megabat            | Megabat            | Pteropus vampyrus               | Mammalia       |\n",
       "| aptMan1  | Brown kiwi         | NA                 | NA                              | Aves           |\n",
       "| dipOrd1  | Kangaroo rat       | NA                 | NA                              | Mammalia       |\n",
       "| oviAri4  | Sheep              | Sheep              | Ovis aries                      | Mammalia       |\n",
       "| allMis1  | American alligator | American alligator | Alligator mississippiensis      | Reptilia       |\n",
       "| canFam5  | Dog                | NA                 | NA                              | Mammalia       |\n",
       "| felCat9  | Cat                | Cat                | Felis catus                     | Mammalia       |\n",
       "| susScr11 | Pig                | Pig                | Sus scrofa                      | Mammalia       |\n",
       "| thaSir1  | Garter snake       | Garter snake       | Thamnophis sirtalis             | Reptilia       |\n",
       "| musFur1  | Ferret             | Ferret             | Mustela putorius furo           | Mammalia       |\n",
       "| calMil1  | Elephant shark     | Elephant shark     | Callorhinchus milii             | Chondrichthyes |\n",
       "| equCab3  | Horse              | NA                 | NA                              | Mammalia       |\n",
       "| speTri2  | Squirrel           | Squirrel           | Ictidomys tridecemlineatus      | Mammalia       |\n",
       "| oryCun2  | Rabbit             | Rabbit             | Oryctolagus cuniculus           | Mammalia       |\n",
       "| cavPor3  | Guinea pig         | NA                 | NA                              | Mammalia       |\n",
       "| tupBel1  | Tree shrew         | Tree shrew         | Tupaia belangeri                | Mammalia       |\n",
       "| taeGut2  | Zebra finch        | Zebra finch        | Taeniopygia guttata             | Aves           |\n",
       "| myoLuc2  | Microbat           | Microbat           | Myotis lucifugus                | Mammalia       |\n",
       "| mm39     | Mouse              | NA                 | NA                              | Mammalia       |\n",
       "| nanPar1  | Tibetan frog       | Tibetan frog       | Nanorana parkeri                | Amphibia       |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| oreNil2       | Nile tilapia             | Nile tilapia         | Oreochromis niloticus | Actinopteri  |\n",
       "| danRer11      | Zebrafish                | Zebrafish            | Danio rerio           | Actinopteri  |\n",
       "| petMar3       | Lamprey                  | NA                   | NA                    | Hyperoartia  |\n",
       "| fr3           | Fugu                     | Fugu                 | Takifugu rubripes     | Actinopteri  |\n",
       "| tetNig2       | Tetraodon                | NA                   | NA                    | Actinopteri  |\n",
       "| geoFor1       | Medium ground finch      | Medium ground finch  | Geospiza fortis       | Aves         |\n",
       "| melGal5       | Turkey                   | NA                   | NA                    | Aves         |\n",
       "| triMan1       | Manatee                  | NA                   | NA                    | Mammalia     |\n",
       "| proCap1       | Rock hyrax               | NA                   | NA                    | Mammalia     |\n",
       "| echTel2       | Tenrec                   | NA                   | NA                    | Mammalia     |\n",
       "| macFas5       | Crab-eating macaque      | NA                   | NA                    | Mammalia     |\n",
       "| chlSab2       | Green Monkey             | Green Monkey         | Chlorocebus sabaeus   | Mammalia     |\n",
       "| rheMac10      | Rhesus                   | Rhesus               | Macaca mulatta        | Mammalia     |\n",
       "| xenLae2       | African clawed frog      | African clawed frog  | Xenopus laevis        | Amphibia     |\n",
       "| xenTro9       | X. tropicalis            | Tropical clawed frog | Xenopus tropicalis    | Amphibia     |\n",
       "| rhiRox1       | Golden snub-nosed monkey | NA                   | NA                    | Mammalia     |\n",
       "| nasLar1       | Proboscis Monkey         | NA                   | NA                    | Mammalia     |\n",
       "| hg38          | Human                    | Human                | Homo sapiens          | Mammalia     |\n",
       "| braFlo1       | Lancelet                 | Lancelet             | Cephalochordata       | Invertebrata |\n",
       "| aplCal1       | Sea Hare                 | Sea Hare             | Aplysia californica   | Invertebrata |\n",
       "| latCha1       | Coelacanth               | Coelacanth           | Latimeria chalumnae   | Actinopteri  |\n",
       "| sarHar1       | Tasmanian devil          | Tasmanian devil      | Sarcophilus harrisii  | Marsupiala   |\n",
       "| monDom5       | Opossum                  | Opossum              | Monodelphis domestica | Marsupiala   |\n",
       "| nomLeu3       | Gibbon                   | Gibbon               | Nomascus leucogenys   | Mammalia     |\n",
       "| gorGor6       | Gorilla                  | Gorilla              | Gorilla gorilla       | Mammalia     |\n",
       "| panTro6       | Chimp                    | Chimp                | Pan troglodytes       | Mammalia     |\n",
       "| panPan2       | Bonobo                   | Bonobo               | Pan paniscus          | Mammalia     |\n",
       "| sorAra2       | Shrew                    | Shrew                | Sorex araneus         | Mammalia     |\n",
       "| macEug2       | Wallaby                  | Wallaby              | Notamacropus eugenii  | Marsupiala   |\n",
       "| AmexG_v6.0-DD | Axolotl                  | Axolotl              | Ambystoma mexicanum   | Amphibia     |\n",
       "\n"
      ],
      "text/plain": [
       "   V1            V2                       V3                  \n",
       "1  vicPac2       Alpaca                   NA                  \n",
       "2  dasNov3       Armadillo                NA                  \n",
       "3  gasAcu1       Stickleback              NA                  \n",
       "4  anoCar2       Lizard                   Lizard              \n",
       "5  melUnd1       Budgerigar               Budgeriger          \n",
       "6  aquChr2       Golden eagle             Golden eagle        \n",
       "7  chrPic1       Painted turtle           Painted turtle      \n",
       "8  calJac4       Marmoset                 NA                  \n",
       "9  choHof1       Sloth                    NA                  \n",
       "10 saiBol1       Squirrel monkey          Squirrel monkey     \n",
       "11 pteVam1       Megabat                  Megabat             \n",
       "12 aptMan1       Brown kiwi               NA                  \n",
       "13 dipOrd1       Kangaroo rat             NA                  \n",
       "14 oviAri4       Sheep                    Sheep               \n",
       "15 allMis1       American alligator       American alligator  \n",
       "16 canFam5       Dog                      NA                  \n",
       "17 felCat9       Cat                      Cat                 \n",
       "18 susScr11      Pig                      Pig                 \n",
       "19 thaSir1       Garter snake             Garter snake        \n",
       "20 musFur1       Ferret                   Ferret              \n",
       "21 calMil1       Elephant shark           Elephant shark      \n",
       "22 equCab3       Horse                    NA                  \n",
       "23 speTri2       Squirrel                 Squirrel            \n",
       "24 oryCun2       Rabbit                   Rabbit              \n",
       "25 cavPor3       Guinea pig               NA                  \n",
       "26 tupBel1       Tree shrew               Tree shrew          \n",
       "27 taeGut2       Zebra finch              Zebra finch         \n",
       "28 myoLuc2       Microbat                 Microbat            \n",
       "29 mm39          Mouse                    NA                  \n",
       "30 nanPar1       Tibetan frog             Tibetan frog        \n",
       "⋮  ⋮             ⋮                        ⋮                   \n",
       "48 oreNil2       Nile tilapia             Nile tilapia        \n",
       "49 danRer11      Zebrafish                Zebrafish           \n",
       "50 petMar3       Lamprey                  NA                  \n",
       "51 fr3           Fugu                     Fugu                \n",
       "52 tetNig2       Tetraodon                NA                  \n",
       "53 geoFor1       Medium ground finch      Medium ground finch \n",
       "54 melGal5       Turkey                   NA                  \n",
       "55 triMan1       Manatee                  NA                  \n",
       "56 proCap1       Rock hyrax               NA                  \n",
       "57 echTel2       Tenrec                   NA                  \n",
       "58 macFas5       Crab-eating macaque      NA                  \n",
       "59 chlSab2       Green Monkey             Green Monkey        \n",
       "60 rheMac10      Rhesus                   Rhesus              \n",
       "61 xenLae2       African clawed frog      African clawed frog \n",
       "62 xenTro9       X. tropicalis            Tropical clawed frog\n",
       "63 rhiRox1       Golden snub-nosed monkey NA                  \n",
       "64 nasLar1       Proboscis Monkey         NA                  \n",
       "65 hg38          Human                    Human               \n",
       "66 braFlo1       Lancelet                 Lancelet            \n",
       "67 aplCal1       Sea Hare                 Sea Hare            \n",
       "68 latCha1       Coelacanth               Coelacanth          \n",
       "69 sarHar1       Tasmanian devil          Tasmanian devil     \n",
       "70 monDom5       Opossum                  Opossum             \n",
       "71 nomLeu3       Gibbon                   Gibbon              \n",
       "72 gorGor6       Gorilla                  Gorilla             \n",
       "73 panTro6       Chimp                    Chimp               \n",
       "74 panPan2       Bonobo                   Bonobo              \n",
       "75 sorAra2       Shrew                    Shrew               \n",
       "76 macEug2       Wallaby                  Wallaby             \n",
       "77 AmexG_v6.0-DD Axolotl                  Axolotl             \n",
       "   V4                              V5            \n",
       "1  NA                              Mammalia      \n",
       "2  NA                              Malacostraca  \n",
       "3  NA                              Actinopteri   \n",
       "4  Anolis carolinensis             Reptilia      \n",
       "5  Melopsittacus undulatus         Aves          \n",
       "6  Aquila chrysaetos               Aves          \n",
       "7  Chrysemys picta                 Reptilia      \n",
       "8  NA                              Mammalia      \n",
       "9  NA                              Mammalia      \n",
       "10 Saimiri boliviensis boliviensis Mammalia      \n",
       "11 Pteropus vampyrus               Mammalia      \n",
       "12 NA                              Aves          \n",
       "13 NA                              Mammalia      \n",
       "14 Ovis aries                      Mammalia      \n",
       "15 Alligator mississippiensis      Reptilia      \n",
       "16 NA                              Mammalia      \n",
       "17 Felis catus                     Mammalia      \n",
       "18 Sus scrofa                      Mammalia      \n",
       "19 Thamnophis sirtalis             Reptilia      \n",
       "20 Mustela putorius furo           Mammalia      \n",
       "21 Callorhinchus milii             Chondrichthyes\n",
       "22 NA                              Mammalia      \n",
       "23 Ictidomys tridecemlineatus      Mammalia      \n",
       "24 Oryctolagus cuniculus           Mammalia      \n",
       "25 NA                              Mammalia      \n",
       "26 Tupaia belangeri                Mammalia      \n",
       "27 Taeniopygia guttata             Aves          \n",
       "28 Myotis lucifugus                Mammalia      \n",
       "29 NA                              Mammalia      \n",
       "30 Nanorana parkeri                Amphibia      \n",
       "⋮  ⋮                               ⋮             \n",
       "48 Oreochromis niloticus           Actinopteri   \n",
       "49 Danio rerio                     Actinopteri   \n",
       "50 NA                              Hyperoartia   \n",
       "51 Takifugu rubripes               Actinopteri   \n",
       "52 NA                              Actinopteri   \n",
       "53 Geospiza fortis                 Aves          \n",
       "54 NA                              Aves          \n",
       "55 NA                              Mammalia      \n",
       "56 NA                              Mammalia      \n",
       "57 NA                              Mammalia      \n",
       "58 NA                              Mammalia      \n",
       "59 Chlorocebus sabaeus             Mammalia      \n",
       "60 Macaca mulatta                  Mammalia      \n",
       "61 Xenopus laevis                  Amphibia      \n",
       "62 Xenopus tropicalis              Amphibia      \n",
       "63 NA                              Mammalia      \n",
       "64 NA                              Mammalia      \n",
       "65 Homo sapiens                    Mammalia      \n",
       "66 Cephalochordata                 Invertebrata  \n",
       "67 Aplysia californica             Invertebrata  \n",
       "68 Latimeria chalumnae             Actinopteri   \n",
       "69 Sarcophilus harrisii            Marsupiala    \n",
       "70 Monodelphis domestica           Marsupiala    \n",
       "71 Nomascus leucogenys             Mammalia      \n",
       "72 Gorilla gorilla                 Mammalia      \n",
       "73 Pan troglodytes                 Mammalia      \n",
       "74 Pan paniscus                    Mammalia      \n",
       "75 Sorex araneus                   Mammalia      \n",
       "76 Notamacropus eugenii            Marsupiala    \n",
       "77 Ambystoma mexicanum             Amphibia      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genomes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8550194e-6443-423b-9b9b-2e15c3a51ca8",
   "metadata": {},
   "outputs": [],
   "source": [
    "genomes_df <-rbind(genomes_df,list(\"AmexG_v6.0-DD\", \"Axolotl\",\"Axolotl\", \"Ambystoma mexicanum\", \"Amphibia\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "36a3c88b-3889-4295-b6e5-fd03d9fa89cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "latin_names <- list(\"vicPac2\" = \"Vicugna pacos\", \"dasNov3\" = \"Dasypus novemcinctus\", \"gasAcu1\" = \"Gasterosteus aculeatus\", \"calJac4\" = \"Callithrix jacchus\", \"choHof1\" = \"Choloepus hoffmanni\",\n",
    "\"aptMan1\" = \"Apteryx australis\", \"dipOrd1\" = \"Dipodomys ordii\", \"canFam5\" = \"Canis lupus familiaris\", \"equCab3\" = \"Equus caballus\", \"cavPor3\" = \"Cavia porcellus\", \"mm39\" = \"Mus musculus\",\n",
    "\"loxAfr3\" = \" Loxodonta africana\", \"neoSch1\" = \"Monachus schauinslandi\", \"eriEur2\" = \"Erinaceus europaeus\", \"bisBis1\" = \"Bison bison bison\", \"criGri1\" = \"Cricetulus griseus\",\n",
    "\"micMur2\" = \"Microcebus murinus\", \"rn7\" = \"Rattus norvegicus\", \"papAnu4\" = \"Papio anubis\", \"strPur2\" = \"Strongylocentrotus purpuratus\", \"ci3\" = \"C. intestinalis\",\n",
    "\"oryLat2\" = \"Oryzias latipes\", \"petMar3\" = \"Petromyzon marinus\", \"tetNig2\" = \"Tetraodon nigroviridis\", \"melGal5\" = \"Meleagris gallopavo\",\"triMan1\" = \" Trichechus manatus latirostris\",\n",
    "\"proCap1\" = \"Procavia capensis\", \"echTel2\" = \"Echinops telfairi\", \"macFas5\" = \"Macaca fascicularis\",  \"rhiRox1\" = \"Rhinopithecus roxellana\", \"nasLar1\" = \"Nasalis larvatus\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "cfaee08f-30c5-40bb-b65c-704f17056163",
   "metadata": {},
   "outputs": [],
   "source": [
    "genomes_df[is.na(V4)]$V4 <- as.character(latin_names[genomes_df[is.na(V4)]$V1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "61bad0b9-e4d8-477b-ab20-50d67d671e3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_summary = data.table()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e97e35a7-5e86-4b9b-b095-835632640e31",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 15 columns (i.e. invalid file). Added 14 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n",
      "Warning message in fread(repeat_file):\n",
      "“Detected 1 column names but the data has 6 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n",
      "Warning message in fread(file.path(genome_dir, genome_id, \"tracks\", ref_file)):\n",
      "“Detected 1 column names but the data has 12 columns (i.e. invalid file). Added 11 extra default column names at the end.”\n",
      "Joining, by = \"genome\"\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for(i in seq_along(genomes_df$V1)){\n",
    "    \n",
    "genome_id <- genomes_df[i,]$V1\n",
    "\n",
    "kmer01 <- as.data.frame(t(read.csv(file.path(genome_dir, genome_id,\"kmer01\"), row.names = 1, sep = \" \")))[\"order\",]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \n",
    "kmer01$genome <- genome_id\n",
    "\n",
    "kmer02 <- as.data.frame(t(read.csv(file.path(genome_dir, genome_id,\"kmer02\"), row.names = 1, sep = \" \"))[\"order\",c(\"CG\"), drop = FALSE])\n",
    "kmer02$genome <- genome_id\n",
    "\n",
    "kmer <- full_join(kmer01, kmer02)\n",
    "\n",
    "kmer$CG_OE <- kmer$CG/(kmer$C*kmer$G)\n",
    "kmer$CG_freq <- kmer$C + kmer$G\n",
    "\n",
    "genome_size <- fread(file.path(genome_dir,genome_id, paste0(genome_id, \".chrom.sizes\")))\n",
    "\n",
    "kmer$total_bp <- sum(genome_size$V2)\n",
    "kmer$N_chr <- NROW(genome_size)\n",
    "\n",
    "annot_files <- list.files(file.path(genome_dir, genome_id, \"tracks\"))\n",
    "repeat_file <- file.path(genome_dir, genome_id, \"tracks\", paste0(genome_id, \"_rmsk.bed\"))\n",
    "if(!file.exists(repeat_file)){\n",
    "    repeat_file <- file.path(genome_dir, genome_id, \"tracks\", paste0(genome_id, \"_simpleRepeat.bed\"))\n",
    "    if(!file.exists(repeat_file)){\n",
    "        \n",
    "        read_repeat_flag <- FALSE\n",
    "    }else{\n",
    "        read_repeat_flag <- TRUE\n",
    "    }\n",
    "}else read_repeat_flag <- TRUE\n",
    "       \n",
    "if(read_repeat_flag){\n",
    "    repeats <- fread(repeat_file)\n",
    "\n",
    "repeats[, width:= V3-V2, ]\n",
    "\n",
    "kmer$repeats_ratio <- sum(repeats$width)/kmer$total_bp\n",
    "}else kmer$repeats_ratio <- NA\n",
    "\n",
    "\n",
    "\n",
    "ref_file <- annot_files[grepl(\"ref\", tolower(annot_files))]\n",
    "\n",
    "gene_annot <- fread(file.path(genome_dir, genome_id, \"tracks\",ref_file))\n",
    "\n",
    "kmer$n_genes_annotated <- NROW(gene_annot)\n",
    "\n",
    "kmer$gene_annotation <- gsub(paste0(genome_id, \"_\"), \"\", gsub(\".bed\", \"\", ref_file))\n",
    "   kmer$common_name <- genomes_df[i,]$V2\n",
    "kmer$latin_name <- genomes_df[i,]$V4 \n",
    "genome_summary <- rbind(kmer, genome_summary)\n",
    "\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "fc0f5604-be1c-4ab4-bc73-d4685b064300",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>A</th><th scope=col>C</th><th scope=col>G</th><th scope=col>T</th><th scope=col>genome</th><th scope=col>CG</th><th scope=col>CG_OE</th><th scope=col>CG_freq</th><th scope=col>total_bp</th><th scope=col>N_chr</th><th scope=col>repeats_ratio</th><th scope=col>n_genes_annotated</th><th scope=col>gene_annotation</th><th scope=col>common_name</th><th scope=col>latin_name</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>0.2700</td><td>0.2300</td><td>0.2300</td><td>0.2700</td><td>AmexG_v6.0-DD</td><td>0.024830</td><td>0.4693762</td><td>0.4600</td><td>26730690244</td><td>    28</td><td>       NA</td><td>1977265</td><td>AmexT_v47-refGenes.gtf</td><td>Axolotl</td><td>Ambystoma mexicanum </td></tr>\n",
       "\t<tr><td>0.3056</td><td>0.1942</td><td>0.1940</td><td>0.3062</td><td>macEug2      </td><td>0.005004</td><td>0.1328209</td><td>0.3882</td><td> 3075184024</td><td>277711</td><td>0.3933385</td><td> 598021</td><td>xenoRefGene           </td><td>Wallaby</td><td>Notamacropus eugenii</td></tr>\n",
       "\t<tr><td>0.2856</td><td>0.2145</td><td>0.2144</td><td>0.2855</td><td>sorAra2      </td><td>0.014470</td><td>0.3146418</td><td>0.4289</td><td> 2423158183</td><td> 12845</td><td>0.2974883</td><td>  23525</td><td>refSeqComposite       </td><td>Shrew  </td><td>Sorex araneus       </td></tr>\n",
       "\t<tr><td>0.2960</td><td>0.2037</td><td>0.2038</td><td>0.2965</td><td>panPan2      </td><td>0.009519</td><td>0.2292958</td><td>0.4075</td><td> 3286643896</td><td> 10274</td><td>0.4209457</td><td>  59345</td><td>refSeqComposite       </td><td>Bonobo </td><td>Pan paniscus        </td></tr>\n",
       "\t<tr><td>0.2962</td><td>0.2035</td><td>0.2036</td><td>0.2967</td><td>panTro6      </td><td>0.010300</td><td>0.2485965</td><td>0.4071</td><td> 3050398082</td><td>  4346</td><td>0.5296771</td><td> 102471</td><td>refSeqComposite       </td><td>Chimp  </td><td>Pan troglodytes     </td></tr>\n",
       "\t<tr><td>0.2966</td><td>0.2032</td><td>0.2032</td><td>0.2969</td><td>gorGor6      </td><td>0.010460</td><td>0.2533286</td><td>0.4064</td><td> 3044872214</td><td>  5486</td><td>0.4955004</td><td>  61950</td><td>refSeqComposite       </td><td>Gorilla</td><td>Gorilla gorilla     </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 15\n",
       "\\begin{tabular}{lllllllllllllll}\n",
       " A & C & G & T & genome & CG & CG\\_OE & CG\\_freq & total\\_bp & N\\_chr & repeats\\_ratio & n\\_genes\\_annotated & gene\\_annotation & common\\_name & latin\\_name\\\\\n",
       " <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <int> & <dbl> & <int> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 0.2700 & 0.2300 & 0.2300 & 0.2700 & AmexG\\_v6.0-DD & 0.024830 & 0.4693762 & 0.4600 & 26730690244 &     28 &        NA & 1977265 & AmexT\\_v47-refGenes.gtf & Axolotl & Ambystoma mexicanum \\\\\n",
       "\t 0.3056 & 0.1942 & 0.1940 & 0.3062 & macEug2       & 0.005004 & 0.1328209 & 0.3882 &  3075184024 & 277711 & 0.3933385 &  598021 & xenoRefGene            & Wallaby & Notamacropus eugenii\\\\\n",
       "\t 0.2856 & 0.2145 & 0.2144 & 0.2855 & sorAra2       & 0.014470 & 0.3146418 & 0.4289 &  2423158183 &  12845 & 0.2974883 &   23525 & refSeqComposite        & Shrew   & Sorex araneus       \\\\\n",
       "\t 0.2960 & 0.2037 & 0.2038 & 0.2965 & panPan2       & 0.009519 & 0.2292958 & 0.4075 &  3286643896 &  10274 & 0.4209457 &   59345 & refSeqComposite        & Bonobo  & Pan paniscus        \\\\\n",
       "\t 0.2962 & 0.2035 & 0.2036 & 0.2967 & panTro6       & 0.010300 & 0.2485965 & 0.4071 &  3050398082 &   4346 & 0.5296771 &  102471 & refSeqComposite        & Chimp   & Pan troglodytes     \\\\\n",
       "\t 0.2966 & 0.2032 & 0.2032 & 0.2969 & gorGor6       & 0.010460 & 0.2533286 & 0.4064 &  3044872214 &   5486 & 0.4955004 &   61950 & refSeqComposite        & Gorilla & Gorilla gorilla     \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 15\n",
       "\n",
       "| A &lt;dbl&gt; | C &lt;dbl&gt; | G &lt;dbl&gt; | T &lt;dbl&gt; | genome &lt;chr&gt; | CG &lt;dbl&gt; | CG_OE &lt;dbl&gt; | CG_freq &lt;dbl&gt; | total_bp &lt;dbl&gt; | N_chr &lt;int&gt; | repeats_ratio &lt;dbl&gt; | n_genes_annotated &lt;int&gt; | gene_annotation &lt;chr&gt; | common_name &lt;chr&gt; | latin_name &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 0.2700 | 0.2300 | 0.2300 | 0.2700 | AmexG_v6.0-DD | 0.024830 | 0.4693762 | 0.4600 | 26730690244 |     28 |        NA | 1977265 | AmexT_v47-refGenes.gtf | Axolotl | Ambystoma mexicanum  |\n",
       "| 0.3056 | 0.1942 | 0.1940 | 0.3062 | macEug2       | 0.005004 | 0.1328209 | 0.3882 |  3075184024 | 277711 | 0.3933385 |  598021 | xenoRefGene            | Wallaby | Notamacropus eugenii |\n",
       "| 0.2856 | 0.2145 | 0.2144 | 0.2855 | sorAra2       | 0.014470 | 0.3146418 | 0.4289 |  2423158183 |  12845 | 0.2974883 |   23525 | refSeqComposite        | Shrew   | Sorex araneus        |\n",
       "| 0.2960 | 0.2037 | 0.2038 | 0.2965 | panPan2       | 0.009519 | 0.2292958 | 0.4075 |  3286643896 |  10274 | 0.4209457 |   59345 | refSeqComposite        | Bonobo  | Pan paniscus         |\n",
       "| 0.2962 | 0.2035 | 0.2036 | 0.2967 | panTro6       | 0.010300 | 0.2485965 | 0.4071 |  3050398082 |   4346 | 0.5296771 |  102471 | refSeqComposite        | Chimp   | Pan troglodytes      |\n",
       "| 0.2966 | 0.2032 | 0.2032 | 0.2969 | gorGor6       | 0.010460 | 0.2533286 | 0.4064 |  3044872214 |   5486 | 0.4955004 |   61950 | refSeqComposite        | Gorilla | Gorilla gorilla      |\n",
       "\n"
      ],
      "text/plain": [
       "  A      C      G      T      genome        CG       CG_OE     CG_freq\n",
       "1 0.2700 0.2300 0.2300 0.2700 AmexG_v6.0-DD 0.024830 0.4693762 0.4600 \n",
       "2 0.3056 0.1942 0.1940 0.3062 macEug2       0.005004 0.1328209 0.3882 \n",
       "3 0.2856 0.2145 0.2144 0.2855 sorAra2       0.014470 0.3146418 0.4289 \n",
       "4 0.2960 0.2037 0.2038 0.2965 panPan2       0.009519 0.2292958 0.4075 \n",
       "5 0.2962 0.2035 0.2036 0.2967 panTro6       0.010300 0.2485965 0.4071 \n",
       "6 0.2966 0.2032 0.2032 0.2969 gorGor6       0.010460 0.2533286 0.4064 \n",
       "  total_bp    N_chr  repeats_ratio n_genes_annotated gene_annotation       \n",
       "1 26730690244     28        NA     1977265           AmexT_v47-refGenes.gtf\n",
       "2  3075184024 277711 0.3933385      598021           xenoRefGene           \n",
       "3  2423158183  12845 0.2974883       23525           refSeqComposite       \n",
       "4  3286643896  10274 0.4209457       59345           refSeqComposite       \n",
       "5  3050398082   4346 0.5296771      102471           refSeqComposite       \n",
       "6  3044872214   5486 0.4955004       61950           refSeqComposite       \n",
       "  common_name latin_name          \n",
       "1 Axolotl     Ambystoma mexicanum \n",
       "2 Wallaby     Notamacropus eugenii\n",
       "3 Shrew       Sorex araneus       \n",
       "4 Bonobo      Pan paniscus        \n",
       "5 Chimp       Pan troglodytes     \n",
       "6 Gorilla     Gorilla gorilla     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(genome_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "83e55ca1-9660-4708-a4c3-43faa3a8063e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>common_name</th><th scope=col>latin_name</th><th scope=col>genome</th><th scope=col>total_bp</th><th scope=col>N_chr</th><th scope=col>A</th><th scope=col>C</th><th scope=col>G</th><th scope=col>T</th><th scope=col>CG</th><th scope=col>CG_OE</th><th scope=col>CG_freq</th><th scope=col>repeats_ratio</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Axolotl</td><td>Ambystoma mexicanum </td><td>AmexG_v6.0-DD</td><td>26730690244</td><td>    28</td><td>0.2700</td><td>0.2300</td><td>0.2300</td><td>0.2700</td><td>0.024830</td><td>0.4693762</td><td>0.4600</td><td>       NA</td></tr>\n",
       "\t<tr><td>Wallaby</td><td>Notamacropus eugenii</td><td>macEug2      </td><td> 3075184024</td><td>277711</td><td>0.3056</td><td>0.1942</td><td>0.1940</td><td>0.3062</td><td>0.005004</td><td>0.1328209</td><td>0.3882</td><td>0.3933385</td></tr>\n",
       "\t<tr><td>Shrew  </td><td>Sorex araneus       </td><td>sorAra2      </td><td> 2423158183</td><td> 12845</td><td>0.2856</td><td>0.2145</td><td>0.2144</td><td>0.2855</td><td>0.014470</td><td>0.3146418</td><td>0.4289</td><td>0.2974883</td></tr>\n",
       "\t<tr><td>Bonobo </td><td>Pan paniscus        </td><td>panPan2      </td><td> 3286643896</td><td> 10274</td><td>0.2960</td><td>0.2037</td><td>0.2038</td><td>0.2965</td><td>0.009519</td><td>0.2292958</td><td>0.4075</td><td>0.4209457</td></tr>\n",
       "\t<tr><td>Chimp  </td><td>Pan troglodytes     </td><td>panTro6      </td><td> 3050398082</td><td>  4346</td><td>0.2962</td><td>0.2035</td><td>0.2036</td><td>0.2967</td><td>0.010300</td><td>0.2485965</td><td>0.4071</td><td>0.5296771</td></tr>\n",
       "\t<tr><td>Gorilla</td><td>Gorilla gorilla     </td><td>gorGor6      </td><td> 3044872214</td><td>  5486</td><td>0.2966</td><td>0.2032</td><td>0.2032</td><td>0.2969</td><td>0.010460</td><td>0.2533286</td><td>0.4064</td><td>0.4955004</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 13\n",
       "\\begin{tabular}{lllllllllllll}\n",
       " common\\_name & latin\\_name & genome & total\\_bp & N\\_chr & A & C & G & T & CG & CG\\_OE & CG\\_freq & repeats\\_ratio\\\\\n",
       " <chr> & <chr> & <chr> & <dbl> & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t Axolotl & Ambystoma mexicanum  & AmexG\\_v6.0-DD & 26730690244 &     28 & 0.2700 & 0.2300 & 0.2300 & 0.2700 & 0.024830 & 0.4693762 & 0.4600 &        NA\\\\\n",
       "\t Wallaby & Notamacropus eugenii & macEug2       &  3075184024 & 277711 & 0.3056 & 0.1942 & 0.1940 & 0.3062 & 0.005004 & 0.1328209 & 0.3882 & 0.3933385\\\\\n",
       "\t Shrew   & Sorex araneus        & sorAra2       &  2423158183 &  12845 & 0.2856 & 0.2145 & 0.2144 & 0.2855 & 0.014470 & 0.3146418 & 0.4289 & 0.2974883\\\\\n",
       "\t Bonobo  & Pan paniscus         & panPan2       &  3286643896 &  10274 & 0.2960 & 0.2037 & 0.2038 & 0.2965 & 0.009519 & 0.2292958 & 0.4075 & 0.4209457\\\\\n",
       "\t Chimp   & Pan troglodytes      & panTro6       &  3050398082 &   4346 & 0.2962 & 0.2035 & 0.2036 & 0.2967 & 0.010300 & 0.2485965 & 0.4071 & 0.5296771\\\\\n",
       "\t Gorilla & Gorilla gorilla      & gorGor6       &  3044872214 &   5486 & 0.2966 & 0.2032 & 0.2032 & 0.2969 & 0.010460 & 0.2533286 & 0.4064 & 0.4955004\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 13\n",
       "\n",
       "| common_name &lt;chr&gt; | latin_name &lt;chr&gt; | genome &lt;chr&gt; | total_bp &lt;dbl&gt; | N_chr &lt;int&gt; | A &lt;dbl&gt; | C &lt;dbl&gt; | G &lt;dbl&gt; | T &lt;dbl&gt; | CG &lt;dbl&gt; | CG_OE &lt;dbl&gt; | CG_freq &lt;dbl&gt; | repeats_ratio &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| Axolotl | Ambystoma mexicanum  | AmexG_v6.0-DD | 26730690244 |     28 | 0.2700 | 0.2300 | 0.2300 | 0.2700 | 0.024830 | 0.4693762 | 0.4600 |        NA |\n",
       "| Wallaby | Notamacropus eugenii | macEug2       |  3075184024 | 277711 | 0.3056 | 0.1942 | 0.1940 | 0.3062 | 0.005004 | 0.1328209 | 0.3882 | 0.3933385 |\n",
       "| Shrew   | Sorex araneus        | sorAra2       |  2423158183 |  12845 | 0.2856 | 0.2145 | 0.2144 | 0.2855 | 0.014470 | 0.3146418 | 0.4289 | 0.2974883 |\n",
       "| Bonobo  | Pan paniscus         | panPan2       |  3286643896 |  10274 | 0.2960 | 0.2037 | 0.2038 | 0.2965 | 0.009519 | 0.2292958 | 0.4075 | 0.4209457 |\n",
       "| Chimp   | Pan troglodytes      | panTro6       |  3050398082 |   4346 | 0.2962 | 0.2035 | 0.2036 | 0.2967 | 0.010300 | 0.2485965 | 0.4071 | 0.5296771 |\n",
       "| Gorilla | Gorilla gorilla      | gorGor6       |  3044872214 |   5486 | 0.2966 | 0.2032 | 0.2032 | 0.2969 | 0.010460 | 0.2533286 | 0.4064 | 0.4955004 |\n",
       "\n"
      ],
      "text/plain": [
       "  common_name latin_name           genome        total_bp    N_chr  A     \n",
       "1 Axolotl     Ambystoma mexicanum  AmexG_v6.0-DD 26730690244     28 0.2700\n",
       "2 Wallaby     Notamacropus eugenii macEug2        3075184024 277711 0.3056\n",
       "3 Shrew       Sorex araneus        sorAra2        2423158183  12845 0.2856\n",
       "4 Bonobo      Pan paniscus         panPan2        3286643896  10274 0.2960\n",
       "5 Chimp       Pan troglodytes      panTro6        3050398082   4346 0.2962\n",
       "6 Gorilla     Gorilla gorilla      gorGor6        3044872214   5486 0.2966\n",
       "  C      G      T      CG       CG_OE     CG_freq repeats_ratio\n",
       "1 0.2300 0.2300 0.2700 0.024830 0.4693762 0.4600         NA    \n",
       "2 0.1942 0.1940 0.3062 0.005004 0.1328209 0.3882  0.3933385    \n",
       "3 0.2145 0.2144 0.2855 0.014470 0.3146418 0.4289  0.2974883    \n",
       "4 0.2037 0.2038 0.2965 0.009519 0.2292958 0.4075  0.4209457    \n",
       "5 0.2035 0.2036 0.2967 0.010300 0.2485965 0.4071  0.5296771    \n",
       "6 0.2032 0.2032 0.2969 0.010460 0.2533286 0.4064  0.4955004    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(genome_summary[,c(14, 15, 5, 9, 10, 1:4, 6:8, 11)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "3e0626f8-1220-4ff2-98ad-cd996cfec84b",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_wt(genome_summary[,c(14, 15, 5, 9, 10, 1:4, 6:8, 11)], file.path(analysis_dir, \"validation\", \"genome_stats.tsv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "53ea314b-a2dc-4a47-bed3-246deec036fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_summary <- fread(file.path(analysis_dir, \"validation\", \"genome_stats.tsv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "04d25aa4-528b-4d78-ac91-547788140390",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>common_name</th><th scope=col>latin_name</th><th scope=col>genome</th><th scope=col>total_bp</th><th scope=col>N_chr</th><th scope=col>A</th><th scope=col>C</th><th scope=col>G</th><th scope=col>T</th><th scope=col>CG</th><th scope=col>CG_OE</th><th scope=col>CG_freq</th><th scope=col>repeats_ratio</th><th scope=col>class</th><th scope=col>download_link</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int64&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Sea Hare       </td><td>Aplysia californica          </td><td>aplCal1</td><td> 715806041</td><td>  8767</td><td>0.300</td><td>0.201</td><td>0.201</td><td>0.299</td><td>0.025</td><td>0.627</td><td>0.402</td><td>0.083</td><td>Invertebrata  </td><td>http://hgdownload.soe.ucsc.edu/goldenPath/aplCal1/bigZips/aplCal1.fa.gz</td></tr>\n",
       "\t<tr><td>Lancelet       </td><td>Cephalochordata              </td><td>braFlo1</td><td> 926386587</td><td>     2</td><td>0.294</td><td>0.206</td><td>0.206</td><td>0.294</td><td>0.028</td><td>0.663</td><td>0.412</td><td>0.063</td><td>Invertebrata  </td><td>http://hgdownload.soe.ucsc.edu/goldenPath/braFlo1/bigZips/braFlo1.fa.gz</td></tr>\n",
       "\t<tr><td>C. intestinalis</td><td>C. intestinalis              </td><td>ci3    </td><td> 115227500</td><td>  1272</td><td>0.322</td><td>0.178</td><td>0.178</td><td>0.322</td><td>0.028</td><td>0.867</td><td>0.357</td><td>0.149</td><td>Invertebrata  </td><td>http://hgdownload.soe.ucsc.edu/goldenPath/ci3/bigZips/ci3.fa.gz        </td></tr>\n",
       "\t<tr><td>S. purpuratus  </td><td>Strongylocentrotus purpuratus</td><td>strPur2</td><td> 907085737</td><td>114223</td><td>0.315</td><td>0.185</td><td>0.185</td><td>0.315</td><td>0.023</td><td>0.676</td><td>0.370</td><td>0.129</td><td>Invertebrata  </td><td>http://hgdownload.soe.ucsc.edu/goldenPath/strPur2/bigZips/strPur2.fa.gz</td></tr>\n",
       "\t<tr><td>Elephant shark </td><td>Callorhinchus milii          </td><td>calMil1</td><td> 974498586</td><td> 21204</td><td>0.288</td><td>0.212</td><td>0.212</td><td>0.288</td><td>0.020</td><td>0.445</td><td>0.423</td><td>0.271</td><td>Chondrichthyes</td><td>http://hgdownload.soe.ucsc.edu/goldenPath/calMil1/bigZips/calMil1.fa.gz</td></tr>\n",
       "\t<tr><td>Coelacanth     </td><td>Latimeria chalumnae          </td><td>latCha1</td><td>2860591921</td><td> 22819</td><td>0.294</td><td>0.206</td><td>0.206</td><td>0.294</td><td>0.012</td><td>0.277</td><td>0.412</td><td>0.071</td><td>Actinopteri   </td><td>http://hgdownload.soe.ucsc.edu/goldenPath/latCha1/bigZips/latCha1.fa.gz</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 15\n",
       "\\begin{tabular}{lllllllllllllll}\n",
       " common\\_name & latin\\_name & genome & total\\_bp & N\\_chr & A & C & G & T & CG & CG\\_OE & CG\\_freq & repeats\\_ratio & class & download\\_link\\\\\n",
       " <chr> & <chr> & <chr> & <int64> & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t Sea Hare        & Aplysia californica           & aplCal1 &  715806041 &   8767 & 0.300 & 0.201 & 0.201 & 0.299 & 0.025 & 0.627 & 0.402 & 0.083 & Invertebrata   & http://hgdownload.soe.ucsc.edu/goldenPath/aplCal1/bigZips/aplCal1.fa.gz\\\\\n",
       "\t Lancelet        & Cephalochordata               & braFlo1 &  926386587 &      2 & 0.294 & 0.206 & 0.206 & 0.294 & 0.028 & 0.663 & 0.412 & 0.063 & Invertebrata   & http://hgdownload.soe.ucsc.edu/goldenPath/braFlo1/bigZips/braFlo1.fa.gz\\\\\n",
       "\t C. intestinalis & C. intestinalis               & ci3     &  115227500 &   1272 & 0.322 & 0.178 & 0.178 & 0.322 & 0.028 & 0.867 & 0.357 & 0.149 & Invertebrata   & http://hgdownload.soe.ucsc.edu/goldenPath/ci3/bigZips/ci3.fa.gz        \\\\\n",
       "\t S. purpuratus   & Strongylocentrotus purpuratus & strPur2 &  907085737 & 114223 & 0.315 & 0.185 & 0.185 & 0.315 & 0.023 & 0.676 & 0.370 & 0.129 & Invertebrata   & http://hgdownload.soe.ucsc.edu/goldenPath/strPur2/bigZips/strPur2.fa.gz\\\\\n",
       "\t Elephant shark  & Callorhinchus milii           & calMil1 &  974498586 &  21204 & 0.288 & 0.212 & 0.212 & 0.288 & 0.020 & 0.445 & 0.423 & 0.271 & Chondrichthyes & http://hgdownload.soe.ucsc.edu/goldenPath/calMil1/bigZips/calMil1.fa.gz\\\\\n",
       "\t Coelacanth      & Latimeria chalumnae           & latCha1 & 2860591921 &  22819 & 0.294 & 0.206 & 0.206 & 0.294 & 0.012 & 0.277 & 0.412 & 0.071 & Actinopteri    & http://hgdownload.soe.ucsc.edu/goldenPath/latCha1/bigZips/latCha1.fa.gz\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 15\n",
       "\n",
       "| common_name &lt;chr&gt; | latin_name &lt;chr&gt; | genome &lt;chr&gt; | total_bp &lt;int64&gt; | N_chr &lt;int&gt; | A &lt;dbl&gt; | C &lt;dbl&gt; | G &lt;dbl&gt; | T &lt;dbl&gt; | CG &lt;dbl&gt; | CG_OE &lt;dbl&gt; | CG_freq &lt;dbl&gt; | repeats_ratio &lt;dbl&gt; | class &lt;chr&gt; | download_link &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| Sea Hare        | Aplysia californica           | aplCal1 |  715806041 |   8767 | 0.300 | 0.201 | 0.201 | 0.299 | 0.025 | 0.627 | 0.402 | 0.083 | Invertebrata   | http://hgdownload.soe.ucsc.edu/goldenPath/aplCal1/bigZips/aplCal1.fa.gz |\n",
       "| Lancelet        | Cephalochordata               | braFlo1 |  926386587 |      2 | 0.294 | 0.206 | 0.206 | 0.294 | 0.028 | 0.663 | 0.412 | 0.063 | Invertebrata   | http://hgdownload.soe.ucsc.edu/goldenPath/braFlo1/bigZips/braFlo1.fa.gz |\n",
       "| C. intestinalis | C. intestinalis               | ci3     |  115227500 |   1272 | 0.322 | 0.178 | 0.178 | 0.322 | 0.028 | 0.867 | 0.357 | 0.149 | Invertebrata   | http://hgdownload.soe.ucsc.edu/goldenPath/ci3/bigZips/ci3.fa.gz         |\n",
       "| S. purpuratus   | Strongylocentrotus purpuratus | strPur2 |  907085737 | 114223 | 0.315 | 0.185 | 0.185 | 0.315 | 0.023 | 0.676 | 0.370 | 0.129 | Invertebrata   | http://hgdownload.soe.ucsc.edu/goldenPath/strPur2/bigZips/strPur2.fa.gz |\n",
       "| Elephant shark  | Callorhinchus milii           | calMil1 |  974498586 |  21204 | 0.288 | 0.212 | 0.212 | 0.288 | 0.020 | 0.445 | 0.423 | 0.271 | Chondrichthyes | http://hgdownload.soe.ucsc.edu/goldenPath/calMil1/bigZips/calMil1.fa.gz |\n",
       "| Coelacanth      | Latimeria chalumnae           | latCha1 | 2860591921 |  22819 | 0.294 | 0.206 | 0.206 | 0.294 | 0.012 | 0.277 | 0.412 | 0.071 | Actinopteri    | http://hgdownload.soe.ucsc.edu/goldenPath/latCha1/bigZips/latCha1.fa.gz |\n",
       "\n"
      ],
      "text/plain": [
       "  common_name     latin_name                    genome  total_bp   N_chr  A    \n",
       "1 Sea Hare        Aplysia californica           aplCal1  715806041   8767 0.300\n",
       "2 Lancelet        Cephalochordata               braFlo1  926386587      2 0.294\n",
       "3 C. intestinalis C. intestinalis               ci3      115227500   1272 0.322\n",
       "4 S. purpuratus   Strongylocentrotus purpuratus strPur2  907085737 114223 0.315\n",
       "5 Elephant shark  Callorhinchus milii           calMil1  974498586  21204 0.288\n",
       "6 Coelacanth      Latimeria chalumnae           latCha1 2860591921  22819 0.294\n",
       "  C     G     T     CG    CG_OE CG_freq repeats_ratio class         \n",
       "1 0.201 0.201 0.299 0.025 0.627 0.402   0.083         Invertebrata  \n",
       "2 0.206 0.206 0.294 0.028 0.663 0.412   0.063         Invertebrata  \n",
       "3 0.178 0.178 0.322 0.028 0.867 0.357   0.149         Invertebrata  \n",
       "4 0.185 0.185 0.315 0.023 0.676 0.370   0.129         Invertebrata  \n",
       "5 0.212 0.212 0.288 0.020 0.445 0.423   0.271         Chondrichthyes\n",
       "6 0.206 0.206 0.294 0.012 0.277 0.412   0.071         Actinopteri   \n",
       "  download_link                                                          \n",
       "1 http://hgdownload.soe.ucsc.edu/goldenPath/aplCal1/bigZips/aplCal1.fa.gz\n",
       "2 http://hgdownload.soe.ucsc.edu/goldenPath/braFlo1/bigZips/braFlo1.fa.gz\n",
       "3 http://hgdownload.soe.ucsc.edu/goldenPath/ci3/bigZips/ci3.fa.gz        \n",
       "4 http://hgdownload.soe.ucsc.edu/goldenPath/strPur2/bigZips/strPur2.fa.gz\n",
       "5 http://hgdownload.soe.ucsc.edu/goldenPath/calMil1/bigZips/calMil1.fa.gz\n",
       "6 http://hgdownload.soe.ucsc.edu/goldenPath/latCha1/bigZips/latCha1.fa.gz"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(genome_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "e0eec563-d8f2-42e9-a992-e2b7a7b59432",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_summary <- left_join(genome_summary, genomes_df[, c(\"V1\", \"V5\")], by = c(\"genome\" =\"V1\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "21e2a380-214e-4750-8868-0009443a40ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>common_name</th><th scope=col>latin_name</th><th scope=col>genome</th><th scope=col>total_bp</th><th scope=col>N_chr</th><th scope=col>A</th><th scope=col>C</th><th scope=col>G</th><th scope=col>T</th><th scope=col>CG</th><th scope=col>CG_OE</th><th scope=col>CG_freq</th><th scope=col>repeats_ratio</th><th scope=col>V5</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int64&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Axolotl</td><td>Ambystoma mexicanum </td><td>AmexG_v6.0-DD</td><td>26730690244</td><td>    28</td><td>0.2700</td><td>0.2300</td><td>0.2300</td><td>0.2700</td><td>0.024830</td><td>0.4693762</td><td>0.4600</td><td>       NA</td><td>Amphibia  </td></tr>\n",
       "\t<tr><td>Wallaby</td><td>Notamacropus eugenii</td><td>macEug2      </td><td> 3075184024</td><td>277711</td><td>0.3056</td><td>0.1942</td><td>0.1940</td><td>0.3062</td><td>0.005004</td><td>0.1328209</td><td>0.3882</td><td>0.3933385</td><td>Marsupiala</td></tr>\n",
       "\t<tr><td>Shrew  </td><td>Sorex araneus       </td><td>sorAra2      </td><td> 2423158183</td><td> 12845</td><td>0.2856</td><td>0.2145</td><td>0.2144</td><td>0.2855</td><td>0.014470</td><td>0.3146418</td><td>0.4289</td><td>0.2974883</td><td>Mammalia  </td></tr>\n",
       "\t<tr><td>Bonobo </td><td>Pan paniscus        </td><td>panPan2      </td><td> 3286643896</td><td> 10274</td><td>0.2960</td><td>0.2037</td><td>0.2038</td><td>0.2965</td><td>0.009519</td><td>0.2292958</td><td>0.4075</td><td>0.4209457</td><td>Mammalia  </td></tr>\n",
       "\t<tr><td>Chimp  </td><td>Pan troglodytes     </td><td>panTro6      </td><td> 3050398082</td><td>  4346</td><td>0.2962</td><td>0.2035</td><td>0.2036</td><td>0.2967</td><td>0.010300</td><td>0.2485965</td><td>0.4071</td><td>0.5296771</td><td>Mammalia  </td></tr>\n",
       "\t<tr><td>Gorilla</td><td>Gorilla gorilla     </td><td>gorGor6      </td><td> 3044872214</td><td>  5486</td><td>0.2966</td><td>0.2032</td><td>0.2032</td><td>0.2969</td><td>0.010460</td><td>0.2533286</td><td>0.4064</td><td>0.4955004</td><td>Mammalia  </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 14\n",
       "\\begin{tabular}{llllllllllllll}\n",
       " common\\_name & latin\\_name & genome & total\\_bp & N\\_chr & A & C & G & T & CG & CG\\_OE & CG\\_freq & repeats\\_ratio & V5\\\\\n",
       " <chr> & <chr> & <chr> & <int64> & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t Axolotl & Ambystoma mexicanum  & AmexG\\_v6.0-DD & 26730690244 &     28 & 0.2700 & 0.2300 & 0.2300 & 0.2700 & 0.024830 & 0.4693762 & 0.4600 &        NA & Amphibia  \\\\\n",
       "\t Wallaby & Notamacropus eugenii & macEug2       &  3075184024 & 277711 & 0.3056 & 0.1942 & 0.1940 & 0.3062 & 0.005004 & 0.1328209 & 0.3882 & 0.3933385 & Marsupiala\\\\\n",
       "\t Shrew   & Sorex araneus        & sorAra2       &  2423158183 &  12845 & 0.2856 & 0.2145 & 0.2144 & 0.2855 & 0.014470 & 0.3146418 & 0.4289 & 0.2974883 & Mammalia  \\\\\n",
       "\t Bonobo  & Pan paniscus         & panPan2       &  3286643896 &  10274 & 0.2960 & 0.2037 & 0.2038 & 0.2965 & 0.009519 & 0.2292958 & 0.4075 & 0.4209457 & Mammalia  \\\\\n",
       "\t Chimp   & Pan troglodytes      & panTro6       &  3050398082 &   4346 & 0.2962 & 0.2035 & 0.2036 & 0.2967 & 0.010300 & 0.2485965 & 0.4071 & 0.5296771 & Mammalia  \\\\\n",
       "\t Gorilla & Gorilla gorilla      & gorGor6       &  3044872214 &   5486 & 0.2966 & 0.2032 & 0.2032 & 0.2969 & 0.010460 & 0.2533286 & 0.4064 & 0.4955004 & Mammalia  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 14\n",
       "\n",
       "| common_name &lt;chr&gt; | latin_name &lt;chr&gt; | genome &lt;chr&gt; | total_bp &lt;int64&gt; | N_chr &lt;int&gt; | A &lt;dbl&gt; | C &lt;dbl&gt; | G &lt;dbl&gt; | T &lt;dbl&gt; | CG &lt;dbl&gt; | CG_OE &lt;dbl&gt; | CG_freq &lt;dbl&gt; | repeats_ratio &lt;dbl&gt; | V5 &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| Axolotl | Ambystoma mexicanum  | AmexG_v6.0-DD | 26730690244 |     28 | 0.2700 | 0.2300 | 0.2300 | 0.2700 | 0.024830 | 0.4693762 | 0.4600 |        NA | Amphibia   |\n",
       "| Wallaby | Notamacropus eugenii | macEug2       |  3075184024 | 277711 | 0.3056 | 0.1942 | 0.1940 | 0.3062 | 0.005004 | 0.1328209 | 0.3882 | 0.3933385 | Marsupiala |\n",
       "| Shrew   | Sorex araneus        | sorAra2       |  2423158183 |  12845 | 0.2856 | 0.2145 | 0.2144 | 0.2855 | 0.014470 | 0.3146418 | 0.4289 | 0.2974883 | Mammalia   |\n",
       "| Bonobo  | Pan paniscus         | panPan2       |  3286643896 |  10274 | 0.2960 | 0.2037 | 0.2038 | 0.2965 | 0.009519 | 0.2292958 | 0.4075 | 0.4209457 | Mammalia   |\n",
       "| Chimp   | Pan troglodytes      | panTro6       |  3050398082 |   4346 | 0.2962 | 0.2035 | 0.2036 | 0.2967 | 0.010300 | 0.2485965 | 0.4071 | 0.5296771 | Mammalia   |\n",
       "| Gorilla | Gorilla gorilla      | gorGor6       |  3044872214 |   5486 | 0.2966 | 0.2032 | 0.2032 | 0.2969 | 0.010460 | 0.2533286 | 0.4064 | 0.4955004 | Mammalia   |\n",
       "\n"
      ],
      "text/plain": [
       "  common_name latin_name           genome        total_bp    N_chr  A     \n",
       "1 Axolotl     Ambystoma mexicanum  AmexG_v6.0-DD 26730690244     28 0.2700\n",
       "2 Wallaby     Notamacropus eugenii macEug2        3075184024 277711 0.3056\n",
       "3 Shrew       Sorex araneus        sorAra2        2423158183  12845 0.2856\n",
       "4 Bonobo      Pan paniscus         panPan2        3286643896  10274 0.2960\n",
       "5 Chimp       Pan troglodytes      panTro6        3050398082   4346 0.2962\n",
       "6 Gorilla     Gorilla gorilla      gorGor6        3044872214   5486 0.2966\n",
       "  C      G      T      CG       CG_OE     CG_freq repeats_ratio V5        \n",
       "1 0.2300 0.2300 0.2700 0.024830 0.4693762 0.4600         NA     Amphibia  \n",
       "2 0.1942 0.1940 0.3062 0.005004 0.1328209 0.3882  0.3933385     Marsupiala\n",
       "3 0.2145 0.2144 0.2855 0.014470 0.3146418 0.4289  0.2974883     Mammalia  \n",
       "4 0.2037 0.2038 0.2965 0.009519 0.2292958 0.4075  0.4209457     Mammalia  \n",
       "5 0.2035 0.2036 0.2967 0.010300 0.2485965 0.4071  0.5296771     Mammalia  \n",
       "6 0.2032 0.2032 0.2969 0.010460 0.2533286 0.4064  0.4955004     Mammalia  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(genome_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "430f5df0-492a-49cd-ac38-3607585caf1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'Amphibia'</li><li>'Marsupiala'</li><li>'Mammalia'</li><li>'Actinopteri'</li><li>'Invertebrata'</li><li>'Aves'</li><li>'Hyperoartia'</li><li>'Ascidiacea'</li><li>'Echinoidea'</li><li>'Chondrichthyes'</li><li>'Reptilia'</li><li>'Malacostraca'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'Amphibia'\n",
       "\\item 'Marsupiala'\n",
       "\\item 'Mammalia'\n",
       "\\item 'Actinopteri'\n",
       "\\item 'Invertebrata'\n",
       "\\item 'Aves'\n",
       "\\item 'Hyperoartia'\n",
       "\\item 'Ascidiacea'\n",
       "\\item 'Echinoidea'\n",
       "\\item 'Chondrichthyes'\n",
       "\\item 'Reptilia'\n",
       "\\item 'Malacostraca'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'Amphibia'\n",
       "2. 'Marsupiala'\n",
       "3. 'Mammalia'\n",
       "4. 'Actinopteri'\n",
       "5. 'Invertebrata'\n",
       "6. 'Aves'\n",
       "7. 'Hyperoartia'\n",
       "8. 'Ascidiacea'\n",
       "9. 'Echinoidea'\n",
       "10. 'Chondrichthyes'\n",
       "11. 'Reptilia'\n",
       "12. 'Malacostraca'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"Amphibia\"       \"Marsupiala\"     \"Mammalia\"       \"Actinopteri\"   \n",
       " [5] \"Invertebrata\"   \"Aves\"           \"Hyperoartia\"    \"Ascidiacea\"    \n",
       " [9] \"Echinoidea\"     \"Chondrichthyes\" \"Reptilia\"       \"Malacostraca\"  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(genome_summary$V5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "efc5069d-668b-449a-8a6b-7af06c334ad5",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(genome_summary)[14] <- \"class\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "d0b596dc-213b-4a11-ad99-6f07fe19e91c",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_summary[class == \"Hyperoartia\",class := \"Jawless vertebrate\",  ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "a492e836-8315-46a3-ac26-4afe240b7d17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>Invertebrata</dt><dd>'#cfcfcf'</dd><dt>Jawless_vertebrate</dt><dd>'#636363'</dd><dt>Chondrichthyes</dt><dd>'#F68383'</dd><dt>Actinopteri</dt><dd>'#EA4142'</dd><dt>Amphibia</dt><dd>'#5AB349'</dd><dt>Reptilia</dt><dd>'#8761AC'</dd><dt>Aves</dt><dd>'#FE9222'</dd><dt>Marsupialia</dt><dd>'#8EBFDA'</dd><dt>Mammalia</dt><dd>'#4892C2'</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[Invertebrata] '\\#cfcfcf'\n",
       "\\item[Jawless\\textbackslash{}\\_vertebrate] '\\#636363'\n",
       "\\item[Chondrichthyes] '\\#F68383'\n",
       "\\item[Actinopteri] '\\#EA4142'\n",
       "\\item[Amphibia] '\\#5AB349'\n",
       "\\item[Reptilia] '\\#8761AC'\n",
       "\\item[Aves] '\\#FE9222'\n",
       "\\item[Marsupialia] '\\#8EBFDA'\n",
       "\\item[Mammalia] '\\#4892C2'\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "Invertebrata\n",
       ":   '#cfcfcf'Jawless_vertebrate\n",
       ":   '#636363'Chondrichthyes\n",
       ":   '#F68383'Actinopteri\n",
       ":   '#EA4142'Amphibia\n",
       ":   '#5AB349'Reptilia\n",
       ":   '#8761AC'Aves\n",
       ":   '#FE9222'Marsupialia\n",
       ":   '#8EBFDA'Mammalia\n",
       ":   '#4892C2'\n",
       "\n"
      ],
      "text/plain": [
       "      Invertebrata Jawless_vertebrate     Chondrichthyes        Actinopteri \n",
       "         \"#cfcfcf\"          \"#636363\"          \"#F68383\"          \"#EA4142\" \n",
       "          Amphibia           Reptilia               Aves        Marsupialia \n",
       "         \"#5AB349\"          \"#8761AC\"          \"#FE9222\"          \"#8EBFDA\" \n",
       "          Mammalia \n",
       "         \"#4892C2\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "class_colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "c8b59bc3-666b-4651-baad-caa758edccf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_summary[class %in% c(\"Echinoidea\",\"Ascidiacea\"), class := \"Invertebrata\",]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "98bcd510-3a34-4679-932f-94275b15bf3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_summary[class == \"Malacostraca\",class := \"Mammalia\",  ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "ac8717fe-06ac-426f-adb3-3655442efc5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_summary[class == \"Marsupiala\",class := \"Marsupialia\",  ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "a4679061-ade3-456e-a1f2-6d12a98e216a",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_summary$class <- factor(genome_summary$class, levels = names(class_colors))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "a905621b-9299-4ec0-bc81-c3c1b8538d44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>common_name</th><th scope=col>latin_name</th><th scope=col>genome</th><th scope=col>total_bp</th><th scope=col>N_chr</th><th scope=col>A</th><th scope=col>C</th><th scope=col>G</th><th scope=col>T</th><th scope=col>CG</th><th scope=col>CG_OE</th><th scope=col>CG_freq</th><th scope=col>repeats_ratio</th><th scope=col>class</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int64&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Sea Hare       </td><td>Aplysia californica          </td><td>aplCal1</td><td> 715806041</td><td>  8767</td><td>0.2995</td><td>0.2008</td><td>0.2007</td><td>0.2991</td><td>0.025</td><td>0.627</td><td>0.402</td><td>0.083</td><td>Invertebrata  </td></tr>\n",
       "\t<tr><td>Lancelet       </td><td>Cephalochordata              </td><td>braFlo1</td><td> 926386587</td><td>     2</td><td>0.2941</td><td>0.2059</td><td>0.2059</td><td>0.2942</td><td>0.028</td><td>0.663</td><td>0.412</td><td>0.063</td><td>Invertebrata  </td></tr>\n",
       "\t<tr><td>C. intestinalis</td><td>C. intestinalis              </td><td>ci3    </td><td> 115227500</td><td>  1272</td><td>0.3215</td><td>0.1783</td><td>0.1785</td><td>0.3218</td><td>0.028</td><td>0.867</td><td>0.357</td><td>0.149</td><td>Invertebrata  </td></tr>\n",
       "\t<tr><td>S. purpuratus  </td><td>Strongylocentrotus purpuratus</td><td>strPur2</td><td> 907085737</td><td>114223</td><td>0.3146</td><td>0.1853</td><td>0.1848</td><td>0.3152</td><td>0.023</td><td>0.676</td><td>0.370</td><td>0.129</td><td>Invertebrata  </td></tr>\n",
       "\t<tr><td>Elephant shark </td><td>Callorhinchus milii          </td><td>calMil1</td><td> 974498586</td><td> 21204</td><td>0.2884</td><td>0.2117</td><td>0.2117</td><td>0.2882</td><td>0.020</td><td>0.445</td><td>0.423</td><td>0.271</td><td>Chondrichthyes</td></tr>\n",
       "\t<tr><td>Coelacanth     </td><td>Latimeria chalumnae          </td><td>latCha1</td><td>2860591921</td><td> 22819</td><td>0.2943</td><td>0.2058</td><td>0.2057</td><td>0.2942</td><td>0.012</td><td>0.277</td><td>0.412</td><td>0.071</td><td>Actinopteri   </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 14\n",
       "\\begin{tabular}{llllllllllllll}\n",
       " common\\_name & latin\\_name & genome & total\\_bp & N\\_chr & A & C & G & T & CG & CG\\_OE & CG\\_freq & repeats\\_ratio & class\\\\\n",
       " <chr> & <chr> & <chr> & <int64> & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <fct>\\\\\n",
       "\\hline\n",
       "\t Sea Hare        & Aplysia californica           & aplCal1 &  715806041 &   8767 & 0.2995 & 0.2008 & 0.2007 & 0.2991 & 0.025 & 0.627 & 0.402 & 0.083 & Invertebrata  \\\\\n",
       "\t Lancelet        & Cephalochordata               & braFlo1 &  926386587 &      2 & 0.2941 & 0.2059 & 0.2059 & 0.2942 & 0.028 & 0.663 & 0.412 & 0.063 & Invertebrata  \\\\\n",
       "\t C. intestinalis & C. intestinalis               & ci3     &  115227500 &   1272 & 0.3215 & 0.1783 & 0.1785 & 0.3218 & 0.028 & 0.867 & 0.357 & 0.149 & Invertebrata  \\\\\n",
       "\t S. purpuratus   & Strongylocentrotus purpuratus & strPur2 &  907085737 & 114223 & 0.3146 & 0.1853 & 0.1848 & 0.3152 & 0.023 & 0.676 & 0.370 & 0.129 & Invertebrata  \\\\\n",
       "\t Elephant shark  & Callorhinchus milii           & calMil1 &  974498586 &  21204 & 0.2884 & 0.2117 & 0.2117 & 0.2882 & 0.020 & 0.445 & 0.423 & 0.271 & Chondrichthyes\\\\\n",
       "\t Coelacanth      & Latimeria chalumnae           & latCha1 & 2860591921 &  22819 & 0.2943 & 0.2058 & 0.2057 & 0.2942 & 0.012 & 0.277 & 0.412 & 0.071 & Actinopteri   \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 14\n",
       "\n",
       "| common_name &lt;chr&gt; | latin_name &lt;chr&gt; | genome &lt;chr&gt; | total_bp &lt;int64&gt; | N_chr &lt;int&gt; | A &lt;dbl&gt; | C &lt;dbl&gt; | G &lt;dbl&gt; | T &lt;dbl&gt; | CG &lt;dbl&gt; | CG_OE &lt;dbl&gt; | CG_freq &lt;dbl&gt; | repeats_ratio &lt;dbl&gt; | class &lt;fct&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| Sea Hare        | Aplysia californica           | aplCal1 |  715806041 |   8767 | 0.2995 | 0.2008 | 0.2007 | 0.2991 | 0.025 | 0.627 | 0.402 | 0.083 | Invertebrata   |\n",
       "| Lancelet        | Cephalochordata               | braFlo1 |  926386587 |      2 | 0.2941 | 0.2059 | 0.2059 | 0.2942 | 0.028 | 0.663 | 0.412 | 0.063 | Invertebrata   |\n",
       "| C. intestinalis | C. intestinalis               | ci3     |  115227500 |   1272 | 0.3215 | 0.1783 | 0.1785 | 0.3218 | 0.028 | 0.867 | 0.357 | 0.149 | Invertebrata   |\n",
       "| S. purpuratus   | Strongylocentrotus purpuratus | strPur2 |  907085737 | 114223 | 0.3146 | 0.1853 | 0.1848 | 0.3152 | 0.023 | 0.676 | 0.370 | 0.129 | Invertebrata   |\n",
       "| Elephant shark  | Callorhinchus milii           | calMil1 |  974498586 |  21204 | 0.2884 | 0.2117 | 0.2117 | 0.2882 | 0.020 | 0.445 | 0.423 | 0.271 | Chondrichthyes |\n",
       "| Coelacanth      | Latimeria chalumnae           | latCha1 | 2860591921 |  22819 | 0.2943 | 0.2058 | 0.2057 | 0.2942 | 0.012 | 0.277 | 0.412 | 0.071 | Actinopteri    |\n",
       "\n"
      ],
      "text/plain": [
       "  common_name     latin_name                    genome  total_bp   N_chr \n",
       "1 Sea Hare        Aplysia californica           aplCal1  715806041   8767\n",
       "2 Lancelet        Cephalochordata               braFlo1  926386587      2\n",
       "3 C. intestinalis C. intestinalis               ci3      115227500   1272\n",
       "4 S. purpuratus   Strongylocentrotus purpuratus strPur2  907085737 114223\n",
       "5 Elephant shark  Callorhinchus milii           calMil1  974498586  21204\n",
       "6 Coelacanth      Latimeria chalumnae           latCha1 2860591921  22819\n",
       "  A      C      G      T      CG    CG_OE CG_freq repeats_ratio class         \n",
       "1 0.2995 0.2008 0.2007 0.2991 0.025 0.627 0.402   0.083         Invertebrata  \n",
       "2 0.2941 0.2059 0.2059 0.2942 0.028 0.663 0.412   0.063         Invertebrata  \n",
       "3 0.3215 0.1783 0.1785 0.3218 0.028 0.867 0.357   0.149         Invertebrata  \n",
       "4 0.3146 0.1853 0.1848 0.3152 0.023 0.676 0.370   0.129         Invertebrata  \n",
       "5 0.2884 0.2117 0.2117 0.2882 0.020 0.445 0.423   0.271         Chondrichthyes\n",
       "6 0.2943 0.2058 0.2057 0.2942 0.012 0.277 0.412   0.071         Actinopteri   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(genome_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "dfc1a81c-0147-4ca4-ace4-5a487d45bffe",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_summary <- genome_summary[order(class)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "ee37beba-b87b-471e-8306-40b6a720e117",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_summary$CG <- round(genome_summary$CG, 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "575f5e22-c3bb-4afe-a247-304a7993071f",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_summary$CG_OE <- round(genome_summary$CG_OE, 3)\n",
    "genome_summary$CG_freq <- round(genome_summary$CG_freq, 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "b1c9b661-4590-4427-b666-53c3ac167df5",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_summary$repeats_ratio <- round(genome_summary$repeats_ratio, 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "0dd31b15-91f1-48e4-a4ae-c415e4954932",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_summary$A <- round(genome_summary$A, 3)\n",
    "genome_summary$C <- round(genome_summary$C, 3)\n",
    "genome_summary$G <- round(genome_summary$G, 3)\n",
    "genome_summary$T <- round(genome_summary$T, 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "7210117b-bb97-4f23-b64c-4de166d9b742",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>common_name</th><th scope=col>latin_name</th><th scope=col>genome</th><th scope=col>total_bp</th><th scope=col>N_chr</th><th scope=col>A</th><th scope=col>C</th><th scope=col>G</th><th scope=col>T</th><th scope=col>CG</th><th scope=col>CG_OE</th><th scope=col>CG_freq</th><th scope=col>repeats_ratio</th><th scope=col>class</th><th scope=col>download_link</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int64&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Sea Hare       </td><td>Aplysia californica          </td><td>aplCal1</td><td> 715806041</td><td>  8767</td><td>0.300</td><td>0.201</td><td>0.201</td><td>0.299</td><td>0.025</td><td>0.627</td><td>0.402</td><td>0.083</td><td>Invertebrata  </td><td>http://hgdownload.soe.ucsc.edu/goldenPath/aplCal1/bigZips/aplCal1.fa.gz</td></tr>\n",
       "\t<tr><td>Lancelet       </td><td>Cephalochordata              </td><td>braFlo1</td><td> 926386587</td><td>     2</td><td>0.294</td><td>0.206</td><td>0.206</td><td>0.294</td><td>0.028</td><td>0.663</td><td>0.412</td><td>0.063</td><td>Invertebrata  </td><td>http://hgdownload.soe.ucsc.edu/goldenPath/braFlo1/bigZips/braFlo1.fa.gz</td></tr>\n",
       "\t<tr><td>C. intestinalis</td><td>C. intestinalis              </td><td>ci3    </td><td> 115227500</td><td>  1272</td><td>0.322</td><td>0.178</td><td>0.178</td><td>0.322</td><td>0.028</td><td>0.867</td><td>0.357</td><td>0.149</td><td>Invertebrata  </td><td>http://hgdownload.soe.ucsc.edu/goldenPath/ci3/bigZips/ci3.fa.gz        </td></tr>\n",
       "\t<tr><td>S. purpuratus  </td><td>Strongylocentrotus purpuratus</td><td>strPur2</td><td> 907085737</td><td>114223</td><td>0.315</td><td>0.185</td><td>0.185</td><td>0.315</td><td>0.023</td><td>0.676</td><td>0.370</td><td>0.129</td><td>Invertebrata  </td><td>http://hgdownload.soe.ucsc.edu/goldenPath/strPur2/bigZips/strPur2.fa.gz</td></tr>\n",
       "\t<tr><td>Elephant shark </td><td>Callorhinchus milii          </td><td>calMil1</td><td> 974498586</td><td> 21204</td><td>0.288</td><td>0.212</td><td>0.212</td><td>0.288</td><td>0.020</td><td>0.445</td><td>0.423</td><td>0.271</td><td>Chondrichthyes</td><td>http://hgdownload.soe.ucsc.edu/goldenPath/calMil1/bigZips/calMil1.fa.gz</td></tr>\n",
       "\t<tr><td>Coelacanth     </td><td>Latimeria chalumnae          </td><td>latCha1</td><td>2860591921</td><td> 22819</td><td>0.294</td><td>0.206</td><td>0.206</td><td>0.294</td><td>0.012</td><td>0.277</td><td>0.412</td><td>0.071</td><td>Actinopteri   </td><td>http://hgdownload.soe.ucsc.edu/goldenPath/latCha1/bigZips/latCha1.fa.gz</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 15\n",
       "\\begin{tabular}{lllllllllllllll}\n",
       " common\\_name & latin\\_name & genome & total\\_bp & N\\_chr & A & C & G & T & CG & CG\\_OE & CG\\_freq & repeats\\_ratio & class & download\\_link\\\\\n",
       " <chr> & <chr> & <chr> & <int64> & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t Sea Hare        & Aplysia californica           & aplCal1 &  715806041 &   8767 & 0.300 & 0.201 & 0.201 & 0.299 & 0.025 & 0.627 & 0.402 & 0.083 & Invertebrata   & http://hgdownload.soe.ucsc.edu/goldenPath/aplCal1/bigZips/aplCal1.fa.gz\\\\\n",
       "\t Lancelet        & Cephalochordata               & braFlo1 &  926386587 &      2 & 0.294 & 0.206 & 0.206 & 0.294 & 0.028 & 0.663 & 0.412 & 0.063 & Invertebrata   & http://hgdownload.soe.ucsc.edu/goldenPath/braFlo1/bigZips/braFlo1.fa.gz\\\\\n",
       "\t C. intestinalis & C. intestinalis               & ci3     &  115227500 &   1272 & 0.322 & 0.178 & 0.178 & 0.322 & 0.028 & 0.867 & 0.357 & 0.149 & Invertebrata   & http://hgdownload.soe.ucsc.edu/goldenPath/ci3/bigZips/ci3.fa.gz        \\\\\n",
       "\t S. purpuratus   & Strongylocentrotus purpuratus & strPur2 &  907085737 & 114223 & 0.315 & 0.185 & 0.185 & 0.315 & 0.023 & 0.676 & 0.370 & 0.129 & Invertebrata   & http://hgdownload.soe.ucsc.edu/goldenPath/strPur2/bigZips/strPur2.fa.gz\\\\\n",
       "\t Elephant shark  & Callorhinchus milii           & calMil1 &  974498586 &  21204 & 0.288 & 0.212 & 0.212 & 0.288 & 0.020 & 0.445 & 0.423 & 0.271 & Chondrichthyes & http://hgdownload.soe.ucsc.edu/goldenPath/calMil1/bigZips/calMil1.fa.gz\\\\\n",
       "\t Coelacanth      & Latimeria chalumnae           & latCha1 & 2860591921 &  22819 & 0.294 & 0.206 & 0.206 & 0.294 & 0.012 & 0.277 & 0.412 & 0.071 & Actinopteri    & http://hgdownload.soe.ucsc.edu/goldenPath/latCha1/bigZips/latCha1.fa.gz\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 15\n",
       "\n",
       "| common_name &lt;chr&gt; | latin_name &lt;chr&gt; | genome &lt;chr&gt; | total_bp &lt;int64&gt; | N_chr &lt;int&gt; | A &lt;dbl&gt; | C &lt;dbl&gt; | G &lt;dbl&gt; | T &lt;dbl&gt; | CG &lt;dbl&gt; | CG_OE &lt;dbl&gt; | CG_freq &lt;dbl&gt; | repeats_ratio &lt;dbl&gt; | class &lt;chr&gt; | download_link &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| Sea Hare        | Aplysia californica           | aplCal1 |  715806041 |   8767 | 0.300 | 0.201 | 0.201 | 0.299 | 0.025 | 0.627 | 0.402 | 0.083 | Invertebrata   | http://hgdownload.soe.ucsc.edu/goldenPath/aplCal1/bigZips/aplCal1.fa.gz |\n",
       "| Lancelet        | Cephalochordata               | braFlo1 |  926386587 |      2 | 0.294 | 0.206 | 0.206 | 0.294 | 0.028 | 0.663 | 0.412 | 0.063 | Invertebrata   | http://hgdownload.soe.ucsc.edu/goldenPath/braFlo1/bigZips/braFlo1.fa.gz |\n",
       "| C. intestinalis | C. intestinalis               | ci3     |  115227500 |   1272 | 0.322 | 0.178 | 0.178 | 0.322 | 0.028 | 0.867 | 0.357 | 0.149 | Invertebrata   | http://hgdownload.soe.ucsc.edu/goldenPath/ci3/bigZips/ci3.fa.gz         |\n",
       "| S. purpuratus   | Strongylocentrotus purpuratus | strPur2 |  907085737 | 114223 | 0.315 | 0.185 | 0.185 | 0.315 | 0.023 | 0.676 | 0.370 | 0.129 | Invertebrata   | http://hgdownload.soe.ucsc.edu/goldenPath/strPur2/bigZips/strPur2.fa.gz |\n",
       "| Elephant shark  | Callorhinchus milii           | calMil1 |  974498586 |  21204 | 0.288 | 0.212 | 0.212 | 0.288 | 0.020 | 0.445 | 0.423 | 0.271 | Chondrichthyes | http://hgdownload.soe.ucsc.edu/goldenPath/calMil1/bigZips/calMil1.fa.gz |\n",
       "| Coelacanth      | Latimeria chalumnae           | latCha1 | 2860591921 |  22819 | 0.294 | 0.206 | 0.206 | 0.294 | 0.012 | 0.277 | 0.412 | 0.071 | Actinopteri    | http://hgdownload.soe.ucsc.edu/goldenPath/latCha1/bigZips/latCha1.fa.gz |\n",
       "\n"
      ],
      "text/plain": [
       "  common_name     latin_name                    genome  total_bp   N_chr  A    \n",
       "1 Sea Hare        Aplysia californica           aplCal1  715806041   8767 0.300\n",
       "2 Lancelet        Cephalochordata               braFlo1  926386587      2 0.294\n",
       "3 C. intestinalis C. intestinalis               ci3      115227500   1272 0.322\n",
       "4 S. purpuratus   Strongylocentrotus purpuratus strPur2  907085737 114223 0.315\n",
       "5 Elephant shark  Callorhinchus milii           calMil1  974498586  21204 0.288\n",
       "6 Coelacanth      Latimeria chalumnae           latCha1 2860591921  22819 0.294\n",
       "  C     G     T     CG    CG_OE CG_freq repeats_ratio class         \n",
       "1 0.201 0.201 0.299 0.025 0.627 0.402   0.083         Invertebrata  \n",
       "2 0.206 0.206 0.294 0.028 0.663 0.412   0.063         Invertebrata  \n",
       "3 0.178 0.178 0.322 0.028 0.867 0.357   0.149         Invertebrata  \n",
       "4 0.185 0.185 0.315 0.023 0.676 0.370   0.129         Invertebrata  \n",
       "5 0.212 0.212 0.288 0.020 0.445 0.423   0.271         Chondrichthyes\n",
       "6 0.206 0.206 0.294 0.012 0.277 0.412   0.071         Actinopteri   \n",
       "  download_link                                                          \n",
       "1 http://hgdownload.soe.ucsc.edu/goldenPath/aplCal1/bigZips/aplCal1.fa.gz\n",
       "2 http://hgdownload.soe.ucsc.edu/goldenPath/braFlo1/bigZips/braFlo1.fa.gz\n",
       "3 http://hgdownload.soe.ucsc.edu/goldenPath/ci3/bigZips/ci3.fa.gz        \n",
       "4 http://hgdownload.soe.ucsc.edu/goldenPath/strPur2/bigZips/strPur2.fa.gz\n",
       "5 http://hgdownload.soe.ucsc.edu/goldenPath/calMil1/bigZips/calMil1.fa.gz\n",
       "6 http://hgdownload.soe.ucsc.edu/goldenPath/latCha1/bigZips/latCha1.fa.gz"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(genome_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f6d5a57d-d38f-4025-a920-a10055b895e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_summary[, download_link:= paste0(\"http://hgdownload.soe.ucsc.edu/goldenPath/\", genome, \"/bigZips/\", genome, \".fa.gz\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c139ca6a-4af1-4093-a6fd-7925ded4bd24",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_summary[common_name== \"Axolotl\",download_link:=\"https://www.axolotl-omics.org/dl/AmexG_v6.0-DD.fa.gz\",]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "d4641034-554f-4231-948c-38c30186c7a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'Common Name'</li><li>'Scientific Name'</li><li>'Genome'</li><li>'Total basepairs'</li><li>'N chromosomes'</li><li>'A freq.'</li><li>'C freq.'</li><li>'G freq.'</li><li>'T freq.'</li><li>'CpG freq.'</li><li>'CpG obs./exp.'</li><li>'CG_freq'</li><li>'Repeats freq.'</li><li>'Taxonomic group'</li><li>'Download link'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'Common Name'\n",
       "\\item 'Scientific Name'\n",
       "\\item 'Genome'\n",
       "\\item 'Total basepairs'\n",
       "\\item 'N chromosomes'\n",
       "\\item 'A freq.'\n",
       "\\item 'C freq.'\n",
       "\\item 'G freq.'\n",
       "\\item 'T freq.'\n",
       "\\item 'CpG freq.'\n",
       "\\item 'CpG obs./exp.'\n",
       "\\item 'CG\\_freq'\n",
       "\\item 'Repeats freq.'\n",
       "\\item 'Taxonomic group'\n",
       "\\item 'Download link'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'Common Name'\n",
       "2. 'Scientific Name'\n",
       "3. 'Genome'\n",
       "4. 'Total basepairs'\n",
       "5. 'N chromosomes'\n",
       "6. 'A freq.'\n",
       "7. 'C freq.'\n",
       "8. 'G freq.'\n",
       "9. 'T freq.'\n",
       "10. 'CpG freq.'\n",
       "11. 'CpG obs./exp.'\n",
       "12. 'CG_freq'\n",
       "13. 'Repeats freq.'\n",
       "14. 'Taxonomic group'\n",
       "15. 'Download link'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"Common Name\"     \"Scientific Name\" \"Genome\"          \"Total basepairs\"\n",
       " [5] \"N chromosomes\"   \"A freq.\"         \"C freq.\"         \"G freq.\"        \n",
       " [9] \"T freq.\"         \"CpG freq.\"       \"CpG obs./exp.\"   \"CG_freq\"        \n",
       "[13] \"Repeats freq.\"   \"Taxonomic group\" \"Download link\"  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colnames(genome_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "f3077457-34e8-471d-9707-3d5f768cf80a",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(genome_summary)[1] <- \"Common Name\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "597070be-ae05-4c72-bcf0-5c161c532750",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(genome_summary)[2] <- \"Scientific Name\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "940d9a06-cd9f-4967-ba63-661751751772",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(genome_summary)[3] <- \"Genome\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "7bd590f9-e2d3-42aa-83ee-392e9b88ac71",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(genome_summary)[6:9] <- paste0(colnames(genome_summary)[6:9], \" freq.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "d45c94d9-4c14-4af5-aa6d-8ca2bc627c4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(genome_summary)[4] <- \"Total basepairs\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "93321039-e6c6-44f1-965d-5792f14070d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(genome_summary)[5] <- \"N chromosomes\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "2ae6c26c-ea87-4860-b1e3-d7a040f67225",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(genome_summary)[10] <- \"CpG freq.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "6a4ff292-24c2-47c9-aaa9-eabe00153e58",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(genome_summary)[11] <- \"CpG obs./exp.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "ae326c00-d979-42fe-9654-38139386ec3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(genome_summary)[13] <- \"Repeats freq.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "0549d241-00bd-4def-8174-7e8407bcd4f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(genome_summary)[14] <- \"Taxonomic group\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "190c624e-9e16-4d3d-a7a6-beab80480727",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(genome_summary)[15] <- \"Download link\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "94ed460f-07b4-4217-b2ac-486c48fca4ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_summary <- genome_summary[, c(1,2,14, 3:11,13,15)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f25facb-4f43-441d-bb61-edb0ab7bcece",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "bff9fd57-e853-4c05-8ab1-d625a8bf1d0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Common Name</th><th scope=col>Scientific Name</th><th scope=col>Taxonomic group</th><th scope=col>Genome</th><th scope=col>Total basepairs</th><th scope=col>N chromosomes</th><th scope=col>A freq.</th><th scope=col>C freq.</th><th scope=col>G freq.</th><th scope=col>T freq.</th><th scope=col>CpG freq.</th><th scope=col>CpG obs./exp.</th><th scope=col>Repeats freq.</th><th scope=col>Download link</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int64&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Sea Hare       </td><td>Aplysia californica          </td><td>Invertebrata  </td><td>aplCal1</td><td> 715806041</td><td>  8767</td><td>0.300</td><td>0.201</td><td>0.201</td><td>0.299</td><td>0.025</td><td>0.627</td><td>0.083</td><td>http://hgdownload.soe.ucsc.edu/goldenPath/aplCal1/bigZips/aplCal1.fa.gz</td></tr>\n",
       "\t<tr><td>Lancelet       </td><td>Cephalochordata              </td><td>Invertebrata  </td><td>braFlo1</td><td> 926386587</td><td>     2</td><td>0.294</td><td>0.206</td><td>0.206</td><td>0.294</td><td>0.028</td><td>0.663</td><td>0.063</td><td>http://hgdownload.soe.ucsc.edu/goldenPath/braFlo1/bigZips/braFlo1.fa.gz</td></tr>\n",
       "\t<tr><td>C. intestinalis</td><td>C. intestinalis              </td><td>Invertebrata  </td><td>ci3    </td><td> 115227500</td><td>  1272</td><td>0.322</td><td>0.178</td><td>0.178</td><td>0.322</td><td>0.028</td><td>0.867</td><td>0.149</td><td>http://hgdownload.soe.ucsc.edu/goldenPath/ci3/bigZips/ci3.fa.gz        </td></tr>\n",
       "\t<tr><td>S. purpuratus  </td><td>Strongylocentrotus purpuratus</td><td>Invertebrata  </td><td>strPur2</td><td> 907085737</td><td>114223</td><td>0.315</td><td>0.185</td><td>0.185</td><td>0.315</td><td>0.023</td><td>0.676</td><td>0.129</td><td>http://hgdownload.soe.ucsc.edu/goldenPath/strPur2/bigZips/strPur2.fa.gz</td></tr>\n",
       "\t<tr><td>Elephant shark </td><td>Callorhinchus milii          </td><td>Chondrichthyes</td><td>calMil1</td><td> 974498586</td><td> 21204</td><td>0.288</td><td>0.212</td><td>0.212</td><td>0.288</td><td>0.020</td><td>0.445</td><td>0.271</td><td>http://hgdownload.soe.ucsc.edu/goldenPath/calMil1/bigZips/calMil1.fa.gz</td></tr>\n",
       "\t<tr><td>Coelacanth     </td><td>Latimeria chalumnae          </td><td>Actinopteri   </td><td>latCha1</td><td>2860591921</td><td> 22819</td><td>0.294</td><td>0.206</td><td>0.206</td><td>0.294</td><td>0.012</td><td>0.277</td><td>0.071</td><td>http://hgdownload.soe.ucsc.edu/goldenPath/latCha1/bigZips/latCha1.fa.gz</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 14\n",
       "\\begin{tabular}{llllllllllllll}\n",
       " Common Name & Scientific Name & Taxonomic group & Genome & Total basepairs & N chromosomes & A freq. & C freq. & G freq. & T freq. & CpG freq. & CpG obs./exp. & Repeats freq. & Download link\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <int64> & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t Sea Hare        & Aplysia californica           & Invertebrata   & aplCal1 &  715806041 &   8767 & 0.300 & 0.201 & 0.201 & 0.299 & 0.025 & 0.627 & 0.083 & http://hgdownload.soe.ucsc.edu/goldenPath/aplCal1/bigZips/aplCal1.fa.gz\\\\\n",
       "\t Lancelet        & Cephalochordata               & Invertebrata   & braFlo1 &  926386587 &      2 & 0.294 & 0.206 & 0.206 & 0.294 & 0.028 & 0.663 & 0.063 & http://hgdownload.soe.ucsc.edu/goldenPath/braFlo1/bigZips/braFlo1.fa.gz\\\\\n",
       "\t C. intestinalis & C. intestinalis               & Invertebrata   & ci3     &  115227500 &   1272 & 0.322 & 0.178 & 0.178 & 0.322 & 0.028 & 0.867 & 0.149 & http://hgdownload.soe.ucsc.edu/goldenPath/ci3/bigZips/ci3.fa.gz        \\\\\n",
       "\t S. purpuratus   & Strongylocentrotus purpuratus & Invertebrata   & strPur2 &  907085737 & 114223 & 0.315 & 0.185 & 0.185 & 0.315 & 0.023 & 0.676 & 0.129 & http://hgdownload.soe.ucsc.edu/goldenPath/strPur2/bigZips/strPur2.fa.gz\\\\\n",
       "\t Elephant shark  & Callorhinchus milii           & Chondrichthyes & calMil1 &  974498586 &  21204 & 0.288 & 0.212 & 0.212 & 0.288 & 0.020 & 0.445 & 0.271 & http://hgdownload.soe.ucsc.edu/goldenPath/calMil1/bigZips/calMil1.fa.gz\\\\\n",
       "\t Coelacanth      & Latimeria chalumnae           & Actinopteri    & latCha1 & 2860591921 &  22819 & 0.294 & 0.206 & 0.206 & 0.294 & 0.012 & 0.277 & 0.071 & http://hgdownload.soe.ucsc.edu/goldenPath/latCha1/bigZips/latCha1.fa.gz\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 14\n",
       "\n",
       "| Common Name &lt;chr&gt; | Scientific Name &lt;chr&gt; | Taxonomic group &lt;chr&gt; | Genome &lt;chr&gt; | Total basepairs &lt;int64&gt; | N chromosomes &lt;int&gt; | A freq. &lt;dbl&gt; | C freq. &lt;dbl&gt; | G freq. &lt;dbl&gt; | T freq. &lt;dbl&gt; | CpG freq. &lt;dbl&gt; | CpG obs./exp. &lt;dbl&gt; | Repeats freq. &lt;dbl&gt; | Download link &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| Sea Hare        | Aplysia californica           | Invertebrata   | aplCal1 |  715806041 |   8767 | 0.300 | 0.201 | 0.201 | 0.299 | 0.025 | 0.627 | 0.083 | http://hgdownload.soe.ucsc.edu/goldenPath/aplCal1/bigZips/aplCal1.fa.gz |\n",
       "| Lancelet        | Cephalochordata               | Invertebrata   | braFlo1 |  926386587 |      2 | 0.294 | 0.206 | 0.206 | 0.294 | 0.028 | 0.663 | 0.063 | http://hgdownload.soe.ucsc.edu/goldenPath/braFlo1/bigZips/braFlo1.fa.gz |\n",
       "| C. intestinalis | C. intestinalis               | Invertebrata   | ci3     |  115227500 |   1272 | 0.322 | 0.178 | 0.178 | 0.322 | 0.028 | 0.867 | 0.149 | http://hgdownload.soe.ucsc.edu/goldenPath/ci3/bigZips/ci3.fa.gz         |\n",
       "| S. purpuratus   | Strongylocentrotus purpuratus | Invertebrata   | strPur2 |  907085737 | 114223 | 0.315 | 0.185 | 0.185 | 0.315 | 0.023 | 0.676 | 0.129 | http://hgdownload.soe.ucsc.edu/goldenPath/strPur2/bigZips/strPur2.fa.gz |\n",
       "| Elephant shark  | Callorhinchus milii           | Chondrichthyes | calMil1 |  974498586 |  21204 | 0.288 | 0.212 | 0.212 | 0.288 | 0.020 | 0.445 | 0.271 | http://hgdownload.soe.ucsc.edu/goldenPath/calMil1/bigZips/calMil1.fa.gz |\n",
       "| Coelacanth      | Latimeria chalumnae           | Actinopteri    | latCha1 | 2860591921 |  22819 | 0.294 | 0.206 | 0.206 | 0.294 | 0.012 | 0.277 | 0.071 | http://hgdownload.soe.ucsc.edu/goldenPath/latCha1/bigZips/latCha1.fa.gz |\n",
       "\n"
      ],
      "text/plain": [
       "  Common Name     Scientific Name               Taxonomic group Genome \n",
       "1 Sea Hare        Aplysia californica           Invertebrata    aplCal1\n",
       "2 Lancelet        Cephalochordata               Invertebrata    braFlo1\n",
       "3 C. intestinalis C. intestinalis               Invertebrata    ci3    \n",
       "4 S. purpuratus   Strongylocentrotus purpuratus Invertebrata    strPur2\n",
       "5 Elephant shark  Callorhinchus milii           Chondrichthyes  calMil1\n",
       "6 Coelacanth      Latimeria chalumnae           Actinopteri     latCha1\n",
       "  Total basepairs N chromosomes A freq. C freq. G freq. T freq. CpG freq.\n",
       "1  715806041        8767        0.300   0.201   0.201   0.299   0.025    \n",
       "2  926386587           2        0.294   0.206   0.206   0.294   0.028    \n",
       "3  115227500        1272        0.322   0.178   0.178   0.322   0.028    \n",
       "4  907085737      114223        0.315   0.185   0.185   0.315   0.023    \n",
       "5  974498586       21204        0.288   0.212   0.212   0.288   0.020    \n",
       "6 2860591921       22819        0.294   0.206   0.206   0.294   0.012    \n",
       "  CpG obs./exp. Repeats freq.\n",
       "1 0.627         0.083        \n",
       "2 0.663         0.063        \n",
       "3 0.867         0.149        \n",
       "4 0.676         0.129        \n",
       "5 0.445         0.271        \n",
       "6 0.277         0.071        \n",
       "  Download link                                                          \n",
       "1 http://hgdownload.soe.ucsc.edu/goldenPath/aplCal1/bigZips/aplCal1.fa.gz\n",
       "2 http://hgdownload.soe.ucsc.edu/goldenPath/braFlo1/bigZips/braFlo1.fa.gz\n",
       "3 http://hgdownload.soe.ucsc.edu/goldenPath/ci3/bigZips/ci3.fa.gz        \n",
       "4 http://hgdownload.soe.ucsc.edu/goldenPath/strPur2/bigZips/strPur2.fa.gz\n",
       "5 http://hgdownload.soe.ucsc.edu/goldenPath/calMil1/bigZips/calMil1.fa.gz\n",
       "6 http://hgdownload.soe.ucsc.edu/goldenPath/latCha1/bigZips/latCha1.fa.gz"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(genome_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "0e66c0d8-4883-4ce2-ba30-f9403412f1cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_wt(genome_summary, file.path(analysis_dir, \"validation\", \"genome_stats.tsv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "fdcaf782-5394-4550-acf6-d6b44c5593b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/binfl/lv71484/droman/DNAmeth500species//results_analysis/validation/genome_stats.tsv'"
      ],
      "text/latex": [
       "'/binfl/lv71484/droman/DNAmeth500species//results\\_analysis/validation/genome\\_stats.tsv'"
      ],
      "text/markdown": [
       "'/binfl/lv71484/droman/DNAmeth500species//results_analysis/validation/genome_stats.tsv'"
      ],
      "text/plain": [
       "[1] \"/binfl/lv71484/droman/DNAmeth500species//results_analysis/validation/genome_stats.tsv\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file.path(analysis_dir, \"validation\", \"genome_stats.tsv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Zoo_R",
   "language": "R",
   "name": "zoo_r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
